{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "26194509-66fb-4e01-b3c6-47c7952a13c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv, find_dotenv\n",
    "\n",
    "load_dotenv(find_dotenv(usecwd=True))\n",
    "\n",
    "base_gg = \"https://gigachat.devices.sberbank.ru/api/v1\"\n",
    "base_gg = \"https://router.huggingface.co/v1\"\n",
    "#\"http://a6k2.dgx:34000/v1\"\n",
    "BASE_URL = os.getenv(\"LITELLM_BASE_URL\", base_gg)\n",
    "\n",
    "BASE_URL = \"https://router.huggingface.co/v1\"\n",
    "\n",
    "auth_key = 'MDE5YmM2MjctOWU5ZS03OWE0LWE0OGYtY2FhMTU4MzYxZDgzOjJmNWUxMzMzLTNlYzMtNGI1Ni04OGZiLTRhMGIxNzFjNzFkMw=='\n",
    "API_KEY = os.getenv(\"LITELLM_API_KEY\", auth_key)\n",
    "#\"qwen3-32b\"\n",
    "# gigachat/GigaChat\n",
    "MODEL_NAME = os.getenv(\"MODEL_NAME\", 'GigaChat')\n",
    "\n",
    "\n",
    "\n",
    "MODEL_NAME_HF ='meta-llama/Llama-3.3-70B-Instruct'\n",
    "\n",
    "MODEL_NAME_HF ='Qwen/Qwen3-32B'\n",
    "MODEL_TEMPERATURE_HF =1\n",
    "HUGGINGFACE_HUB_TOKEN ='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'\n",
    "\n",
    "\n",
    "os.environ[\"OPENAI_API_KEY\"] = HUGGINGFACE_HUB_TOKEN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e14f07d4-cfbd-4129-8f11-5bf5c74d31cf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\user\\anaconda3\\envs\\llms\\Lib\\site-packages\\langchain_core\\_api\\deprecation.py:26: UserWarning: Core Pydantic V1 functionality isn't compatible with Python 3.14 or greater.\n",
      "  from pydantic.v1.fields import FieldInfo as FieldInfoV1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Consider using the pymupdf_layout package for a greatly improved page layout analysis.\n"
     ]
    }
   ],
   "source": [
    "from langchain_community.document_loaders import PyMuPDFLoader\n",
    "from pydantic import BaseModel, Field\n",
    "from langchain_core.output_parsers import PydanticOutputParser, StrOutputParser, BaseOutputParser\n",
    "from textwrap import dedent\n",
    "\n",
    "from langchain_openai import ChatOpenAI\n",
    "from langchain_core.messages import HumanMessage\n",
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "import httpx\n",
    "import json\n",
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "from langchain_pymupdf4llm import PyMuPDF4LLMLoader, PyMuPDF4LLMParser\n",
    "import pymupdf4llm\n",
    "\n",
    "from trialmind.pubmed import pmid2papers, PubmedAPIWrapper, pmid2biocxml, parse_bioc_xml\n",
    "from trialmind.api import StudyCharacteristicsExtraction, ScreeningCriteriaGeneration\n",
    "from trialmind.retrievers import split_text_into_chunks\n",
    "\n",
    "\n",
    "import time\n",
    "import extract\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5301994a-a15c-47ce-b6b8-ce0ceae9e67b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### clinical trials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7eb1ed4e-88ad-454d-a841-868d8d03925b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Colorectal cancer',\n",
       " ['Crizotinib',\n",
       "  'Lorlatinib',\n",
       "  'Necitumumab',\n",
       "  'Nimotuzumab',\n",
       "  'Panitumumab',\n",
       "  'Cetuximab',\n",
       "  'Brigatinib',\n",
       "  'Ramucirumab',\n",
       "  'Pomalidomide',\n",
       "  'Toremifene',\n",
       "  'Denosumab',\n",
       "  'Duvelisib',\n",
       "  'Inavolitib',\n",
       "  'Bevacizumab',\n",
       "  'Anastrozole',\n",
       "  'Exemestane'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"docs/CC_215_L00_DL_fusion_fixed_oncobox_ru.pdf\"\n",
    "fin_condition, treatements_eng = extract.info_from_doc(file_path)\n",
    "fin_condition, treatements_eng"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3c68437-54be-4611-9aa0-d7f6a379a924",
   "metadata": {},
   "source": [
    "1. загрузить док\n",
    "2. взять 1-ую страницу; рег.выражениями взять диагноз\n",
    "-- с ЛЛМ перевести на англ.\n",
    "3. найти страницу с таблицей; или взять таблицы с PyMuPDFLoader\n",
    "4. извлечь таблицу"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "e04c59f8-3cbd-4491-9154-56edbc434e12",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Crizotinib OR Lorlatinib OR Necitumumab OR Nimotuzumab OR Panitumumab OR Cetuximab OR Brigatinib OR Ramucirumab OR Pomalidomide OR Toremifene OR Denosumab OR Duvelisib OR Inavolitib OR Bevacizumab OR Anastrozole OR Exemestane'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "' OR '.join(treatements_eng)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6b530143-823c-495a-8e2a-7039031a23b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0b55cf5c-e883-4a42-ad8f-08633873be8f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_studies = extract.get_clinicaltrials(fin_condition, \n",
    "                                      #' OR '.join(treatements_eng), \n",
    "                                       treatements_eng[0],  \n",
    "                                      max_studies=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30556d38-abf1-4810-a15e-46d3eef966dc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>results</th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>status</th>\n",
       "      <th>conditions</th>\n",
       "      <th>study_type</th>\n",
       "      <th>phases</th>\n",
       "      <th>interventions</th>\n",
       "      <th>outcomes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT02465060</td>\n",
       "      <td>Targeted Therapy Directed by Genetic Testing i...</td>\n",
       "      <td>ACTIVE_NOT_RECRUITING</td>\n",
       "      <td>[Advanced Lymphoma, Advanced Malignant Solid N...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[adavosertib, afatinib, afatinib dimaleate, bi...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT03947385</td>\n",
       "      <td>Study of IDE196 in Patients With Solid Tumors ...</td>\n",
       "      <td>RECRUITING</td>\n",
       "      <td>[Metastatic Uveal Melanoma, Cutaneous Melanoma...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE1, PHASE2]</td>\n",
       "      <td>[ide196, binimetinib, crizotinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT02568267</td>\n",
       "      <td>Basket Study of Entrectinib (RXDX-101) for the...</td>\n",
       "      <td>ACTIVE_NOT_RECRUITING</td>\n",
       "      <td>[Breast Cancer, Cholangiocarcinoma, Colorectal...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[entrectinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT06214793</td>\n",
       "      <td>Taletrectinib in Previously Treated Metastatic...</td>\n",
       "      <td>SUSPENDED</td>\n",
       "      <td>[Breast Cancer, Metastatic Breast Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[taletrectinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT03792568</td>\n",
       "      <td>ALK Inhibitor in Metastatic Colorectal Cancer ...</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>[Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT05725200</td>\n",
       "      <td>Study to Investigate Outcome of Individualized...</td>\n",
       "      <td>RECRUITING</td>\n",
       "      <td>[Metastatic Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[alectinib, cetuximab, crizotinib, dasatinib, ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>NCT02510001</td>\n",
       "      <td>MEK and MET Inhibition in Colorectal Cancer</td>\n",
       "      <td>COMPLETED</td>\n",
       "      <td>[Solid Tumor, Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE1]</td>\n",
       "      <td>[pf-02341066, pd-0325901, binimetinib]</td>\n",
       "      <td>[{'type': 'PRIMARY', 'title': 'Maximal Tolerat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   results           id                                              title  \\\n",
       "0    False  NCT02465060  Targeted Therapy Directed by Genetic Testing i...   \n",
       "1    False  NCT03947385  Study of IDE196 in Patients With Solid Tumors ...   \n",
       "2    False  NCT02568267  Basket Study of Entrectinib (RXDX-101) for the...   \n",
       "3    False  NCT06214793  Taletrectinib in Previously Treated Metastatic...   \n",
       "4    False  NCT03792568  ALK Inhibitor in Metastatic Colorectal Cancer ...   \n",
       "5    False  NCT05725200  Study to Investigate Outcome of Individualized...   \n",
       "6     True  NCT02510001        MEK and MET Inhibition in Colorectal Cancer   \n",
       "\n",
       "                  status                                         conditions  \\\n",
       "0  ACTIVE_NOT_RECRUITING  [Advanced Lymphoma, Advanced Malignant Solid N...   \n",
       "1             RECRUITING  [Metastatic Uveal Melanoma, Cutaneous Melanoma...   \n",
       "2  ACTIVE_NOT_RECRUITING  [Breast Cancer, Cholangiocarcinoma, Colorectal...   \n",
       "3              SUSPENDED          [Breast Cancer, Metastatic Breast Cancer]   \n",
       "4                UNKNOWN                                [Colorectal Cancer]   \n",
       "5             RECRUITING                     [Metastatic Colorectal Cancer]   \n",
       "6              COMPLETED                   [Solid Tumor, Colorectal Cancer]   \n",
       "\n",
       "       study_type            phases  \\\n",
       "0  INTERVENTIONAL          [PHASE2]   \n",
       "1  INTERVENTIONAL  [PHASE1, PHASE2]   \n",
       "2  INTERVENTIONAL          [PHASE2]   \n",
       "3  INTERVENTIONAL          [PHASE2]   \n",
       "4  INTERVENTIONAL              [NA]   \n",
       "5  INTERVENTIONAL          [PHASE2]   \n",
       "6  INTERVENTIONAL          [PHASE1]   \n",
       "\n",
       "                                       interventions  \\\n",
       "0  [adavosertib, afatinib, afatinib dimaleate, bi...   \n",
       "1                  [ide196, binimetinib, crizotinib]   \n",
       "2                                      [entrectinib]   \n",
       "3                                    [taletrectinib]   \n",
       "4                                                 []   \n",
       "5  [alectinib, cetuximab, crizotinib, dasatinib, ...   \n",
       "6             [pf-02341066, pd-0325901, binimetinib]   \n",
       "\n",
       "                                            outcomes  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2                                                NaN  \n",
       "3                                                NaN  \n",
       "4                                                NaN  \n",
       "5                                                NaN  \n",
       "6  [{'type': 'PRIMARY', 'title': 'Maximal Tolerat...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_studies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37f1ccfb-f682-457e-bf61-1bed5aad563c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### outcome results\n",
    "Взяли все результаты, отдельно сохранили испытания с результатами, и з них отбираем рез-ты только с unit==participants\n",
    "\n",
    "filter by interventional study type?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "c5e73eef-7cac-4115-93c7-34ba0ee46532",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'type': 'PRIMARY',\n",
       "  'title': 'Clinical Response to Binimetinib Combined With PF-02341066',\n",
       "  'description': 'To investigate response to treatment with RPII dose of Binimetinib with Crizotinib (PF-02341066), in patients with a) RASMT CRC or b) RASWT/cMET mut amplified CRC or c) RASWT/c-MET over-expressed CRC, as defined by stable, partially or completely responding disease, per Response Evaluation Criteria In Solid Tumors Criteria (RECIST v1.1) for target lesions and assessed by CT: Complete Response (CR), Disappearance of all target lesions; Partial Response (PR), \\\\>=30% decrease in the sum of the longest diameter of target lesions; Stable Disease (SD), Neither sufficient shrinkage to qualify for PR nor sufficient increase (\\\\>=20%) to qualify for Progressive Disease; Overall Response (OR) = CR + PR + SD',\n",
       "  'populationDescription': 'Evaluable patients for the primary outcome are those patients who complete a response assessment after cycle 1 of treatment, or who progress early on treatment. 30 of the 36 recruited patients had a response assessment after cycle 1 or progressed early on treatment, and hence these 30 patients are evaluable for the primary analysis.',\n",
       "  'reportingStatus': 'POSTED',\n",
       "  'paramType': 'COUNT_OF_PARTICIPANTS',\n",
       "  'unitOfMeasure': 'Participants',\n",
       "  'timeFrame': 'Dose Expansion phase: change from baseline and up to 12 months.',\n",
       "  'groups': [{'id': 'OG000',\n",
       "    'title': 'Dose Expansion Phase',\n",
       "    'description': 'Binimetinib 30mg BD interval dose administration Days 1-21 every 28 days PF-02341066 (Crizotinib) 250mg OD Days 1-28 continuously Dosage determined following the recommended Phase II dose identification in the dose escalation phase.\\n\\nPF-02341066: PF-02341066 (Crizotinib) 250mg OD or 200mg BD or 250mg BD Days 1-28 continuously\\n\\nBinimetinib: Binimetinib 30mg or 45 mg BD either continuously or Days 1-21 every 28 days'}],\n",
       "  'denoms': [{'units': 'Participants',\n",
       "    'counts': [{'groupId': 'OG000', 'value': '30'}]}],\n",
       "  'classes': [{'categories': [{'title': 'Stable Disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '7'}]},\n",
       "     {'title': 'Progressive Disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '22'}]},\n",
       "     {'title': 'Early death from malignant disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '1'}]}]}]}]"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outcome_participants = [obj for item_main in results_studies for obj in item_main['resultsSection']['outcomeMeasuresModule'\n",
    "                                           ]['outcomeMeasures'\n",
    "                                            ] if(obj.get('unitOfMeasure','') == 'Participants')] \n",
    "outcome_participants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9303a48c-9dd0-418b-8c9f-3708b4f9148a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "db6360d3-76fa-4607-8764-0a705af4464c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<think>\n",
      "\n",
      "</think>\n",
      "\n",
      "Here is a concise summary of the efficiency statistics for the clinical response to Binimetinib combined with PF-02341066 (Crizotinib) in the Dose Expansion Phase:\n",
      "\n",
      "- **Total Evaluable Patients**: 30 participants.\n",
      "- **Stable Disease (SD)**: 7 patients (23.3%).\n",
      "- **Progressive Disease (PD)**: 22 patients (73.3%).\n",
      "- **Early Death from Malignant Disease**: 1 patient (3.3%).\n",
      "\n",
      "These outcomes were assessed using RECIST v1.1 criteria after cycle 1 of treatment.\n"
     ]
    }
   ],
   "source": [
    "prompt2 = ChatPromptTemplate.from_messages([\n",
    "    (\"system\", \"You are a helpful medical assistant. Extract information about the efficiency statistics from each outcome measure. Give a concise description with percentages /no_think\"),\n",
    "    (\"human\", \"{text}\"),\n",
    "    # (\"ai\", \"Good.\")\n",
    "])\n",
    "\n",
    "chain2 = prompt2 | llm_reasoning\n",
    "response_res = await chain2.ainvoke({\"text\": outcome_participants[0]})\n",
    "print(response_res.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84f5d879-2920-4ec3-bac4-5ae64fb55099",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e706ee64-c65f-40a9-8606-8e71c97fa5d6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### results in a given format?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f1d0545-8778-43fb-9af8-da1730e79453",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class Info(BaseModel):\n",
    "    diagnose: str = Field(..., description=\"База клинических испытаний\")\n",
    "    treatments: str = Field(..., description=\"Высшие позиции рейтинга таргетных препаратов\")\n",
    "\n",
    "parser = PydanticOutputParser(pydantic_object=Info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d5b0edc-bb8f-4fcd-b443-36745ad7eaf7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "prompt = ChatPromptTemplate.from_messages([\n",
    "    (\n",
    "        \"system\", \n",
    "        dedent(\n",
    "            \"\"\"\n",
    "            You are a helpful assistant. \n",
    "            Your goal is to extract related information from the text provided by the user.\n",
    "            Return the information in the following JSON-format.\n",
    "            {fmt}\n",
    "            \"\"\"\n",
    "        )\n",
    "    ),\n",
    "    (\n",
    "        \"human\", \n",
    "        dedent(\n",
    "            \"\"\"\n",
    "            Text: {text}\"\n",
    "            \"\"\"\n",
    "        )\n",
    "    ),\n",
    "])\n",
    "chain = (\n",
    "    prompt \n",
    "    | llm_reasoning \n",
    "    | parser\n",
    ")\n",
    "\n",
    "\n",
    "response = await chain.ainvoke({\"text\": text, \"fmt\": parser.get_format_instructions()})\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d963b7db-0c01-4b5f-af74-d7d566ae8a54",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ceab5c26-765a-420c-82a6-8bfbacdb36f6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['41268440', '41213063', '40526090', '40211189', '40369167'],\n",
       " 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=(Colorectal cancer)+AND+(Crizotinib)&retmax=5&retmode=json&api_key=6892c4129cef143ff92d11533848d2e0d908',\n",
       " '67')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "search_api = PubmedAPIWrapper()\n",
    "# page_size is the max number of records to return!!!! not pages!\n",
    "tmp_inputs = {\n",
    "        \"page_size\": 5,\n",
    "        \"keyword_map\": {'conditions':[fin_condition], \n",
    "                        'treatments':[treatements_eng[0]]\n",
    "                       },\n",
    "        \"keywords\": {\n",
    "            \"OPERATOR\": 'AND'\n",
    "        }\n",
    "}\n",
    "pubmed_api_key= '6892c4129cef143ff92d11533848d2e0d908'#os.getenv(\"PUBMED_API_KEY\", '')\n",
    "time.sleep(.5)\n",
    "response = search_api.build_search_query_and_get_pmid(tmp_inputs, \n",
    "                                                      api_key=pubmed_api_key)\n",
    "response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a54a4f2c-d723-4a4a-a1b1-77eb50eeaea8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Year</th>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th>Title</th>\n",
       "      <th>Publication Type</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>41268440</td>\n",
       "      <td>Frontiers in pharmacology</td>\n",
       "      <td>2025</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Case Report: HGF and NF1 mutations as putative...</td>\n",
       "      <td>Case Reports, Journal Article</td>\n",
       "      <td>Zhitao Chen, Shan Luo, Yangjun Gu, Qiyong Li</td>\n",
       "      <td>Hepatocellular carcinoma (HCC) is a highly agg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>41213063</td>\n",
       "      <td>Journal of environmental pathology, toxicology...</td>\n",
       "      <td>2025</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Radiotherapy-Related Autophagy Genes Predict P...</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Mao Mao, Yichao Zhuang, Yan Chen</td>\n",
       "      <td>Numerous genes have been associated with color...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40526090</td>\n",
       "      <td>Oncotarget</td>\n",
       "      <td>2025</td>\n",
       "      <td>Jun</td>\n",
       "      <td>17</td>\n",
       "      <td>Case Report WIN-MTB-2023001 WIN International ...</td>\n",
       "      <td>Journal Article, Case Reports</td>\n",
       "      <td>Alberto Hernando-Calvo, Razelle Kurzrock, Nadi...</td>\n",
       "      <td>Heavily pretreated metastatic colorectal cance...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>40211189</td>\n",
       "      <td>BMC cancer</td>\n",
       "      <td>2025</td>\n",
       "      <td>Apr</td>\n",
       "      <td>10</td>\n",
       "      <td>A Phase Ia/b study of MEK1/2 inhibitor binimet...</td>\n",
       "      <td>Journal Article, Clinical Trial, Phase I</td>\n",
       "      <td>Francesca Aroldi, Elena Elez, Thierry André, G...</td>\n",
       "      <td>Targeting RAS mutant (MT) colorectal cancer (C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40369167</td>\n",
       "      <td>Cell death and differentiation</td>\n",
       "      <td>2025</td>\n",
       "      <td>Nov</td>\n",
       "      <td></td>\n",
       "      <td>ARID1A loss enhances sensitivity to c-MET inhi...</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Xu Zhang, Zihuan Wang, Yilin He, Kejin Wang, C...</td>\n",
       "      <td>ARID1A, a subunit of the SWI/SNF chromatin-rem...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                            Journal  Year Month  \\\n",
       "0  41268440                          Frontiers in pharmacology  2025         \n",
       "1  41213063  Journal of environmental pathology, toxicology...  2025         \n",
       "2  40526090                                         Oncotarget  2025   Jun   \n",
       "3  40211189                                         BMC cancer  2025   Apr   \n",
       "4  40369167                     Cell death and differentiation  2025   Nov   \n",
       "\n",
       "  Day                                              Title  \\\n",
       "0      Case Report: HGF and NF1 mutations as putative...   \n",
       "1      Radiotherapy-Related Autophagy Genes Predict P...   \n",
       "2  17  Case Report WIN-MTB-2023001 WIN International ...   \n",
       "3  10  A Phase Ia/b study of MEK1/2 inhibitor binimet...   \n",
       "4      ARID1A loss enhances sensitivity to c-MET inhi...   \n",
       "\n",
       "                           Publication Type  \\\n",
       "0             Case Reports, Journal Article   \n",
       "1                           Journal Article   \n",
       "2             Journal Article, Case Reports   \n",
       "3  Journal Article, Clinical Trial, Phase I   \n",
       "4                           Journal Article   \n",
       "\n",
       "                                             Authors  \\\n",
       "0       Zhitao Chen, Shan Luo, Yangjun Gu, Qiyong Li   \n",
       "1                   Mao Mao, Yichao Zhuang, Yan Chen   \n",
       "2  Alberto Hernando-Calvo, Razelle Kurzrock, Nadi...   \n",
       "3  Francesca Aroldi, Elena Elez, Thierry André, G...   \n",
       "4  Xu Zhang, Zihuan Wang, Yilin He, Kejin Wang, C...   \n",
       "\n",
       "                                            Abstract  \n",
       "0  Hepatocellular carcinoma (HCC) is a highly agg...  \n",
       "1  Numerous genes have been associated with color...  \n",
       "2  Heavily pretreated metastatic colorectal cance...  \n",
       "3  Targeting RAS mutant (MT) colorectal cancer (C...  \n",
       "4  ARID1A, a subunit of the SWI/SNF chromatin-rem...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_papers = pmid2papers(pmid_list=response[0], api_key=pubmed_api_key)\n",
    "df_papers[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "03862166-a563-44e1-9682-bdb2437c8075",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['41268440', '41213063', '40526090', '40211189', '40369167']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_papers[0].PMID.values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dd4f5664-3983-4bbb-af9f-7f460b87e5e3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Colorectal cancer',\n",
       " ['Crizotinib',\n",
       "  'Lorlatinib',\n",
       "  'Necitumumab',\n",
       "  'Nimotuzumab',\n",
       "  'Panitumumab',\n",
       "  'Cetuximab',\n",
       "  'Brigatinib',\n",
       "  'Ramucirumab',\n",
       "  'Pomalidomide',\n",
       "  'Toremifene',\n",
       "  'Denosumab',\n",
       "  'Duvelisib',\n",
       "  'Inavolitib',\n",
       "  'Bevacizumab',\n",
       "  'Anastrozole',\n",
       "  'Exemestane'])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_condition, treatements_eng"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f15bedd0-7b97-4696-ac71-2b9ddc4ef0d6",
   "metadata": {},
   "source": [
    "### extract results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "365da54d-32aa-445c-b27b-0b6b2ae80cf6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fin_condition = 'Colorectal cancer'\n",
    "treatements_eng =  ['Crizotinib','Lorlatinib',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "66ffb1de-672b-4c10-acfc-bcade11f8c7d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\user\\anaconda3\\envs\\llms\\Lib\\site-packages\\urllib3\\connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host 'ngw.devices.sberbank.ru'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import uuid\n",
    "\n",
    "url = \"https://ngw.devices.sberbank.ru:9443/api/v2/oauth\"\n",
    "# Создадим идентификатор UUID (36 знаков)\n",
    "rq_uid = str(uuid.uuid4())\n",
    "payload={\n",
    "  'scope': 'GIGACHAT_API_PERS'\n",
    "}\n",
    "headers = {\n",
    "  'Content-Type': 'application/x-www-form-urlencoded',\n",
    "  'Accept': 'application/json',\n",
    "  'RqUID': rq_uid,\n",
    "  'Authorization': f'Basic {auth_key}'\n",
    "}\n",
    "\n",
    "response = requests.request(\"POST\", url, headers=headers, data=payload, verify=False)\n",
    "API_KEY = response.json()['access_token']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0d84ade6-0fcc-4e33-bd67-0aa7ec613b46",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "eyJjdHkiOiJqd3QiLCJlbmMiOiJBMjU2Q0JDLUhTNTEyIiwiYWxnIjoiUlNBLU9BRVAtMjU2In0.dFbPi_ghjzxdARH4qWEIBaezISVD81PtO_E77tUKHGVboF1rT4d1E0hStNt9NU7W5_VAgSuN6L-lF3xtQyhix0Gm3Nnl8lSDD5s5ZETs3jNJB-IxQBen-b2SJqaY7nmmr0j_-G9BfQekQS7CzazA27xbwFjWvYXKiLQO_kwo1B3LPicSG7YPlL8qOiGXEK8gu75OoDUpzcdpEWo2xfiNuDzvnuVIVA9WQlzD5C0pUNkJosdjLx-KjWRbCwCJKszielVpfkKWNBX9fwwhujfFBCYGzTODXwRpuBDXyJu5lyy_Zm1SKv9AgcVj1xFriBAfZZmv-CtBj91eNIvjiwTziQ.W9K55PqgpKV_XHqXxduPVg.n_AlC1aYGE55aipXraRZrLxoZkMWzPlORkm3CvNK8NV3pFDQG0QX_sIQdXBXbfNAVoL-nZEOPOqbdu5ER5vLEf7dbvrOwcu606Y6RpfKWjz6PX7-0OkBCPdsecejqnEhxx1MTZfnLUAqTsr2gb0vkdI3rDTutxYAvUbGPVAIkUR87d54LtPrU0jvdK0J8MD2rcoy8riCy8atqSjMbo1z7U-lnKhg4TDS8SyxSuLs6_jNTa9I90l1E_G5jqZbOLUR3dn-tgqUMImg7NCjlI41_fhzy2gk48ZWBu9YMLb6FXV_jHDUapI6DmhXAcLnDKFYir984LFFKXOeHVU56vgOk0NO5XWzrQy25cVDqUDfjahWpZtwvpv8BXK8px5EeHcjscFdgay5i_aUHy-DwS1NvOCoJebRquBkJalHvrPr7WiB0Hp50CqdNs87sUzoctKR8Q0n3_Y80y0eyUWXqbLjJQUhXFUUKRPqy94kIJXOs47xkADCKCkHQOz9lfr580b_URCO9MDNwR3I_C8LYlqaJNivGff2F_nsOTjWxpKuHdfaEaaQ8rnJBZMQ-FfXsj7E7EA0hhTbUdOvBJxkkUDAj8f4qEluZNjSV4RrKMsZANHcpGm_QXgPlrPJl7UQvz4yEPxJcEzT5xvhgc0EoEHs-afF_KJ6AGca_ecI6iPQ_hV2nwRFSjehYZBQ3PF3zvY2UW5t25GQCrd8equxR5cWJe-HKk_2GJ0FS1_biwB0jtE.6c9jkw4ctaKmPSmBxHPT3NFqNTvj_bJOHcqjQcTLT48\n"
     ]
    }
   ],
   "source": [
    "print(API_KEY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "720ecef2-268b-4b8e-8241-96b44d26d0bf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from openai import OpenAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4e632a26-387b-4c21-b67c-5408ce327484",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "BASE_URL_gc = \"https://gigachat.devices.sberbank.ru/api/v1\"\n",
    "API_KEY_gc = auth_key\n",
    "MODEL_NAME_gc = 'GigaChat'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "be0895aa-5d22-4c08-b855-21edebeff92f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChatCompletion(id=None, choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='**Свет вечерний тихо струится,**\\n**Словно нежный шелковый плат,**\\n**И березы шепчут дружно:**\\n**«Будем вместе до рассвета!»**\\n\\n**Звёзды в небе ярко мерцают,**\\n**Тихо ночь укроет сады,**\\n**А душа моя мечтает —**\\n**О любви и о звездах вновь.»**', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None))], created=1770216233, model='GigaChat:2.0.28.2', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=83, prompt_tokens=20, total_tokens=103, completion_tokens_details=None, prompt_tokens_details=None, precached_prompt_tokens=2))"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BASE_URL = \"https://gigachat.devices.sberbank.ru/api/v1\"\n",
    "HUGGINGFACE_HUB_TOKEN='eyJjdHkiOiJqd3QiLCJlbmMiOiJBMjU2Q0JDLUhTNTEyIiwiYWxnIjoiUlNBLU9BRVAtMjU2In0.YRJxY_Gpcr0OTl66zliP9vq2lCwdsOVUk2oxOfE_FwBI-teifDVHdPHP5OiU2lk4r8jwA9KDTSnRLFYPRc7zTey5F2068fN_nA4wHwY2BC5YMENX-0RyFpsdKAQDtt0bAk_NuYu89n5MMCrFfTMW5Gfx73Y8-T3RSguM9jcNZUJIZIAlhhPD7v8OVjIhh8gL9AD1F94VvYHmdv75DQC63-JYcnqUyZx25fWd1A3SNyN6ebHqceSKsvRtCZ-TyKCdoxJDmyVC0p8GIAZfYirlXY7Rs8Be3sp_nbWiz7LlfvHv1yI6zDRJ_9K9n0aWm7NajeEPmF0QYVHmOXS6pWNdZQ.kAV9Ys5xdV2uUzzDxHyZSQ.fr70dO8KNVeVTKsnPkBRadNVB13OA5PT5ss3jr2baLZgkF2S60WYNPG-VU5A42XrVdzrmZIaok6iOGDo_PMoKw8UQ-IzAjuHehaZaS_3pODnUq0POeRLDzWBaHS6iUk0TIOEOrZZGtgbNfHwOYqbQ6kXKRaKnT7wLyMa3l-tZ8yWFK-dHB-Ocr9bhk5qliO5z7kSbOupOETwHgdMXbda_nDpK5RELaEEaYwDjzdRuGooTQwP9BcqMs4EUHbE2WFQZWp0TarL5RJx7EgD6oWBiyknRP7X3vpvnOj6XofyQ4Wl-JAFraOKqbk9cvezmb6Y6PDu7MXQs8OXsjHXBQy_Rhe3rzKx9DWZS-GgGE-vD2VG6N7WoVM1mgzJnmXLff9Bw4WjjeKGDTmnmA3-cMUFRzi5R4wmVRQ8mA473-JCAG1sqHzyEiaizJnfTifAZOHLkMUZe-3DFKsMXFxmCCtU40va5K0Nx_dxTgR1nmv2ZxPGi-U_rdSwRKA2qQwKU1ErFId9WI2tSq19pqpXh5R4tCkruewR7uEeG9FLsiHm_585gTvWLT_H-cV-Bw-miJ8gwJRUnaqoQ_9T_yrIKSLvlFe8oMOH6qJJ5eDS9IUfdu2tFffVo8GlD8XY01-qw8IOwS5N4IMda1wpz7cpaPf8m789vcNVCDWnDMISXjWROdAC9yUF1oGFlnat-RxiflqXkuLMKM2H6Sx2n51yxLSaP9HknE0QSjuZjO0oWdEhjuo.Ew6MEu9pgPk5Skl1Q708YXu4qsfAEC8mYQlT86rc2GI'\n",
    "\n",
    "openai_client = OpenAI(\n",
    "    base_url=BASE_URL,\n",
    "    api_key=HUGGINGFACE_HUB_TOKEN,\n",
    "    http_client=httpx.Client(http2=True, verify=False)\n",
    ")\n",
    "\n",
    "response = openai_client.chat.completions.create(\n",
    "        model=MODEL_NAME_gc,\n",
    "        messages=[{\"role\": \"user\", \"content\": \"You are a helpful assistant. Create a short poem\"}],  # Ensure messages is a list\n",
    "        temperature=0\n",
    "    )\n",
    "response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "729e1c01-3636-4799-a4d7-c9fe9f303603",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "URL('https://gigachat.devices.sberbank.ru/api/v1/')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "openai_client.base_url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f7987e16-3579-4cf1-b9a0-ef66a903aeb9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://gigachat.devices.sberbank.ru/api/v1/ eyJjdHkiOiJqd3QiLCJlbmMiOiJBMjU2Q0JDLUhTNTEyIiwiYWxnIjoiUlNBLU9BRVAtMjU2In0.YRJxY_Gpcr0OTl66zliP9vq2lCwdsOVUk2oxOfE_FwBI-teifDVHdPHP5OiU2lk4r8jwA9KDTSnRLFYPRc7zTey5F2068fN_nA4wHwY2BC5YMENX-0RyFpsdKAQDtt0bAk_NuYu89n5MMCrFfTMW5Gfx73Y8-T3RSguM9jcNZUJIZIAlhhPD7v8OVjIhh8gL9AD1F94VvYHmdv75DQC63-JYcnqUyZx25fWd1A3SNyN6ebHqceSKsvRtCZ-TyKCdoxJDmyVC0p8GIAZfYirlXY7Rs8Be3sp_nbWiz7LlfvHv1yI6zDRJ_9K9n0aWm7NajeEPmF0QYVHmOXS6pWNdZQ.kAV9Ys5xdV2uUzzDxHyZSQ.fr70dO8KNVeVTKsnPkBRadNVB13OA5PT5ss3jr2baLZgkF2S60WYNPG-VU5A42XrVdzrmZIaok6iOGDo_PMoKw8UQ-IzAjuHehaZaS_3pODnUq0POeRLDzWBaHS6iUk0TIOEOrZZGtgbNfHwOYqbQ6kXKRaKnT7wLyMa3l-tZ8yWFK-dHB-Ocr9bhk5qliO5z7kSbOupOETwHgdMXbda_nDpK5RELaEEaYwDjzdRuGooTQwP9BcqMs4EUHbE2WFQZWp0TarL5RJx7EgD6oWBiyknRP7X3vpvnOj6XofyQ4Wl-JAFraOKqbk9cvezmb6Y6PDu7MXQs8OXsjHXBQy_Rhe3rzKx9DWZS-GgGE-vD2VG6N7WoVM1mgzJnmXLff9Bw4WjjeKGDTmnmA3-cMUFRzi5R4wmVRQ8mA473-JCAG1sqHzyEiaizJnfTifAZOHLkMUZe-3DFKsMXFxmCCtU40va5K0Nx_dxTgR1nmv2ZxPGi-U_rdSwRKA2qQwKU1ErFId9WI2tSq19pqpXh5R4tCkruewR7uEeG9FLsiHm_585gTvWLT_H-cV-Bw-miJ8gwJRUnaqoQ_9T_yrIKSLvlFe8oMOH6qJJ5eDS9IUfdu2tFffVo8GlD8XY01-qw8IOwS5N4IMda1wpz7cpaPf8m789vcNVCDWnDMISXjWROdAC9yUF1oGFlnat-RxiflqXkuLMKM2H6Sx2n51yxLSaP9HknE0QSjuZjO0oWdEhjuo.Ew6MEu9pgPk5Skl1Q708YXu4qsfAEC8mYQlT86rc2GI\n",
      "{\n",
      "  \"ELIGIBILITY_ANALYSIS\": [\n",
      "    // Eligibility Criteria Rationales\n",
      "    \"Patients must have colorectal cancer — Studies involving other types of cancers would not provide relevant data.\",\n",
      "    \"Studies must include patients receiving Crizotinib as an intervention — Inclusion of other treatments would dilute the focus on Crizotinib's effects.\",\n",
      "    \"Comparisons must involve standard therapies or placebos — Comparing Crizotinib against no therapy or alternative interventions unrelated to Crizotinib would introduce bias.\",\n",
      "    \"Outcomes must measure efficacy and safety outcomes such as response rate, progression-free survival, overall survival, toxicity, etc. — Other endpoints may not be directly related to the therapeutic effect of Crizotinib.\",\n",
      "    \"Study designs must be randomized controlled trials or observational studies with appropriate controls — Non-randomized studies might lead to confounding factors affecting results.\"\n",
      "  ],\n",
      "  \"TITLE_CRITERIA\": [\n",
      "    // Binary Title-Based Questions\n",
      "    \"Does the title mention 'colorectal cancer'?\",\n",
      "    \"Does the title mention 'Crizotinib'?\",\n",
      "    \"Is there any indication in the title that the study compares Crizotinib to another treatment?\"\n",
      "  ],\n",
      "  \"CONTENT_CRITERIA\": [\n",
      "    // Binary Content-Based Questions\n",
      "    \"Do the methods section describe randomization and allocation concealment?\",\n",
      "    \"Are the outcomes reported including response rates, progression-free survival, or overall survival?\",\n",
      "    \"Were the participants treated with Crizotinib alone or in combination with other agents?\"\n",
      "  ]\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'title_criteria': [], 'content_criteria': [], 'eligibility_analysis': []}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test the apis\n",
    "\n",
    "api = ScreeningCriteriaGeneration()\n",
    "ecs = api.run(\n",
    "    population = f\"Patients with {fin_condition} undergoing treatment with {treatements_eng[0]}\",\n",
    "    intervention = f\"{treatements_eng[0]}\",\n",
    "    comparator = \"\",\n",
    "    outcome = \"\",\n",
    "    llm = MODEL_NAME_gc,#'Qwen/Qwen3-32B',\n",
    "    num_title_criteria=3,\n",
    "    num_abstract_criteria=3,\n",
    "    thinking=False,\n",
    ")\n",
    "ecs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ab49e6f3-3d7b-477e-ba04-0ff770ff2424",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def extract_json(input_text):\n",
    "    # Pattern to match content between ```json and ```\n",
    "    json_pattern = r\"```json\\n([\\s\\S]*?)\\n```\"\n",
    "    match = re.search(json_pattern, input_text)\n",
    "\n",
    "    if match:\n",
    "        # Extract JSON content between ```json and ```\n",
    "        return match.group(1)\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        # Pattern to match content between {{ and }}\n",
    "        curly_pattern = r\"\\{\\{([\\s\\S]*?)\\}\\}\"\n",
    "        match = re.search(curly_pattern, input_text)\n",
    "        if match:\n",
    "            output = match.group(1)\n",
    "            \n",
    "        else:\n",
    "            # Attempt to parse the entire input as JSON\n",
    "            try:\n",
    "                json.loads(input_text)\n",
    "                # If no exception is raised, the entire input is valid JSON\n",
    "                return input_text\n",
    "            except json.JSONDecodeError:\n",
    "                # Input is not valid JSON\n",
    "                return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "97f5dc1a-29d5-48dd-a9f6-2eb49d30abb7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "mes = '''{\"ELIGIBILITY_ANALYSIS\": [\n",
    "    // Eligibility Criteria Rationales\n",
    "    \"Patients must have colorectal cancer — Studies involving other types of cancers would not provide relevant data.\",\n",
    "    \"Studies must include patients receiving Crizotinib as an intervention — Inclusion of other treatments would dilute the focus on Crizotinib's effects.\",\n",
    "    \"Comparisons must involve standard therapies or placebos — Comparing Crizotinib against no therapy or alternative interventions unrelated to Crizotinib would introduce bias.\",\n",
    "    \"Outcomes must measure efficacy and safety outcomes such as response rate, progression-free survival, overall survival, toxicity, etc. — Other endpoints may not be directly related to the therapeutic effect of Crizotinib.\",\n",
    "    \"Study designs must be randomized controlled trials or observational studies with appropriate controls — Non-randomized studies might lead to confounding factors affecting results.\"\n",
    "  ],\n",
    "  \"TITLE_CRITERIA\": [\n",
    "    // Binary Title-Based Questions\n",
    "    \"Does the title mention 'colorectal cancer'?\",\n",
    "    \"Does the title mention 'Crizotinib'?\",\n",
    "    \"Is there any indication in the title that the study compares Crizotinib to another treatment?\"\n",
    "  ],\n",
    "  \"CONTENT_CRITERIA\": [\n",
    "    // Binary Content-Based Questions\n",
    "    \"Do the methods section describe randomization and allocation concealment?\",\n",
    "    \"Are the outcomes reported including response rates, progression-free survival, or overall survival?\",\n",
    "    \"Were the participants treated with Crizotinib alone or in combination with other agents?\"\n",
    "  ]}'''\n",
    "extract_json(mes.strip('\\n'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7de5f47e-3823-4352-8d8d-f2f94415ca59",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'{\"ELIGIBILITY_ANALYSIS\": [\\n    // Eligibility Criteria Rationales\\n    \"Patients must have colorectal cancer — Studies involving other types of cancers would not provide relevant data.\",\\n    \"Studies must include patients receiving Crizotinib as an intervention — Inclusion of other treatments would dilute the focus on Crizotinib\\'s effects.\",\\n    \"Comparisons must involve standard therapies or placebos — Comparing Crizotinib against no therapy or alternative interventions unrelated to Crizotinib would introduce bias.\",\\n    \"Outcomes must measure efficacy and safety outcomes such as response rate, progression-free survival, overall survival, toxicity, etc. — Other endpoints may not be directly related to the therapeutic effect of Crizotinib.\",\\n    \"Study designs must be randomized controlled trials or observational studies with appropriate controls — Non-randomized studies might lead to confounding factors affecting results.\"\\n  ],\\n  \"TITLE_CRITERIA\": [\\n    // Binary Title-Based Questions\\n    \"Does the title mention \\'colorectal cancer\\'?\",\\n    \"Does the title mention \\'Crizotinib\\'?\",\\n    \"Is there any indication in the title that the study compares Crizotinib to another treatment?\"\\n  ],\\n  \"CONTENT_CRITERIA\": [\\n    // Binary Content-Based Questions\\n    \"Do the methods section describe randomization and allocation concealment?\",\\n    \"Are the outcomes reported including response rates, progression-free survival, or overall survival?\",\\n    \"Were the participants treated with Crizotinib alone or in combination with other agents?\"\\n  ]}'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mes.strip('ELIGIBILITY_ANALYSIS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b1a1479d-5d31-4489-af35-d066463df0e8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "JSONDecodeError",
     "evalue": "Expecting value: line 2 column 5 (char 31)",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mJSONDecodeError\u001b[39m                           Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[33]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mjson\u001b[49m\u001b[43m.\u001b[49m\u001b[43mloads\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmes\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~\\anaconda3\\envs\\llms\\Lib\\json\\__init__.py:352\u001b[39m, in \u001b[36mloads\u001b[39m\u001b[34m(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001b[39m\n\u001b[32m    347\u001b[39m     s = s.decode(detect_encoding(s), \u001b[33m'\u001b[39m\u001b[33msurrogatepass\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m    349\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m (\u001b[38;5;28mcls\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m object_hook \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m\n\u001b[32m    350\u001b[39m         parse_int \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m parse_float \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m\n\u001b[32m    351\u001b[39m         parse_constant \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m object_pairs_hook \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m kw):\n\u001b[32m--> \u001b[39m\u001b[32m352\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_default_decoder\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdecode\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    353\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mcls\u001b[39m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    354\u001b[39m     \u001b[38;5;28mcls\u001b[39m = JSONDecoder\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~\\anaconda3\\envs\\llms\\Lib\\json\\decoder.py:345\u001b[39m, in \u001b[36mJSONDecoder.decode\u001b[39m\u001b[34m(self, s, _w)\u001b[39m\n\u001b[32m    340\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mdecode\u001b[39m(\u001b[38;5;28mself\u001b[39m, s, _w=WHITESPACE.match):\n\u001b[32m    341\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"Return the Python representation of ``s`` (a ``str`` instance\u001b[39;00m\n\u001b[32m    342\u001b[39m \u001b[33;03m    containing a JSON document).\u001b[39;00m\n\u001b[32m    343\u001b[39m \n\u001b[32m    344\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m345\u001b[39m     obj, end = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mraw_decode\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43midx\u001b[49m\u001b[43m=\u001b[49m\u001b[43m_w\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mend\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    346\u001b[39m     end = _w(s, end).end()\n\u001b[32m    347\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m end != \u001b[38;5;28mlen\u001b[39m(s):\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~\\anaconda3\\envs\\llms\\Lib\\json\\decoder.py:363\u001b[39m, in \u001b[36mJSONDecoder.raw_decode\u001b[39m\u001b[34m(self, s, idx)\u001b[39m\n\u001b[32m    361\u001b[39m     obj, end = \u001b[38;5;28mself\u001b[39m.scan_once(s, idx)\n\u001b[32m    362\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[32m--> \u001b[39m\u001b[32m363\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m JSONDecodeError(\u001b[33m\"\u001b[39m\u001b[33mExpecting value\u001b[39m\u001b[33m\"\u001b[39m, s, err.value) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m    364\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m obj, end\n",
      "\u001b[31mJSONDecodeError\u001b[39m: Expecting value: line 2 column 5 (char 31)"
     ]
    }
   ],
   "source": [
    "json.loads(mes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "cf038950-f68c-425a-a677-d24436e971e3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'{\\n    \"ELIGIBILITY_ANALYSIS\": [\\n        \"1. Studies must include patients with cancer undergoing treatment with immune checkpoint inhibitors (ICIs). Rationale: This ensures the population aligns with the PICO element \\'P\\' and focuses on the specific group of interest.\",\\n        \"2. Studies must report on the occurrence of immune-related adverse events (irAEs). Rationale: This aligns with the PICO element \\'I\\' and is essential for comparing outcomes between patients who experienced irAEs and those who did not.\",\\n        \"3. Studies must include a comparison group of patients who did not develop irAEs. Rationale: This is necessary to meet the PICO element \\'C\\' and to allow for a meaningful comparison of outcomes.\",\\n        \"4. Studies must report on overall survival (OS) and/or progression-free survival (PFS). Rationale: These are the key outcomes (PICO element \\'O\\') of the meta-analysis and are critical for assessing the impact of irAEs on clinical outcomes.\",\\n        \"5. Studies must be observational or interventional clinical studies (e.g., cohort, case-control, or randomized controlled trials). Rationale: These study designs provide reliable data on associations between irAEs and survival outcomes, which is essential for a clinical meta-analysis.\"\\n    ],\\n    \"TITLE_CRITERIA\": [\\n        \"Does the title mention \\'immune checkpoint inhibitors\\' or \\'ICIs\\'?\",\\n        \"Does the title include terms related to \\'adverse events\\' or \\'immune-related adverse events\\'?\",\\n        \"Does the title mention \\'survival\\' or \\'prognosis\\' in the context of cancer treatment?\"\\n    ],\\n    \"CONTENT_CRITERIA\": [\\n        \"Does the study explicitly define and report on immune-related adverse events (irAEs)?\",\\n        \"Does the study provide data on overall survival (OS) and/or progression-free survival (PFS) for both patients with and without irAEs?\",\\n        \"Does the study compare outcomes between patients who experienced irAEs and those who did not?\"\\n    ]\\n}'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from trialmind.api import extract_json\n",
    "extract_json('''{\n",
    "    \"ELIGIBILITY_ANALYSIS\": [\n",
    "        \"1. Studies must include patients with cancer undergoing treatment with immune checkpoint inhibitors (ICIs). Rationale: This ensures the population aligns with the PICO element 'P' and focuses on the specific group of interest.\",\n",
    "        \"2. Studies must report on the occurrence of immune-related adverse events (irAEs). Rationale: This aligns with the PICO element 'I' and is essential for comparing outcomes between patients who experienced irAEs and those who did not.\",\n",
    "        \"3. Studies must include a comparison group of patients who did not develop irAEs. Rationale: This is necessary to meet the PICO element 'C' and to allow for a meaningful comparison of outcomes.\",\n",
    "        \"4. Studies must report on overall survival (OS) and/or progression-free survival (PFS). Rationale: These are the key outcomes (PICO element 'O') of the meta-analysis and are critical for assessing the impact of irAEs on clinical outcomes.\",\n",
    "        \"5. Studies must be observational or interventional clinical studies (e.g., cohort, case-control, or randomized controlled trials). Rationale: These study designs provide reliable data on associations between irAEs and survival outcomes, which is essential for a clinical meta-analysis.\"\n",
    "    ],\n",
    "    \"TITLE_CRITERIA\": [\n",
    "        \"Does the title mention 'immune checkpoint inhibitors' or 'ICIs'?\",\n",
    "        \"Does the title include terms related to 'adverse events' or 'immune-related adverse events'?\",\n",
    "        \"Does the title mention 'survival' or 'prognosis' in the context of cancer treatment?\"\n",
    "    ],\n",
    "    \"CONTENT_CRITERIA\": [\n",
    "        \"Does the study explicitly define and report on immune-related adverse events (irAEs)?\",\n",
    "        \"Does the study provide data on overall survival (OS) and/or progression-free survival (PFS) for both patients with and without irAEs?\",\n",
    "        \"Does the study compare outcomes between patients who experienced irAEs and those who did not?\"\n",
    "    ]\n",
    "}''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fa3705c-65ed-4004-9c4d-a7006ca417c7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b46fe7b-fe13-49e7-b0db-7a3dbab33f44",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c69ff7c4-4e91-4afc-bdae-ec1fbf094012",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/41268440/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/41213063/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40526090/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40211189/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40369167/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n"
     ]
    }
   ],
   "source": [
    "pmid_list = df_papers[0].PMID.values.tolist()\n",
    "#['41213063',#'26451310',]\n",
    "\n",
    "res = pmid2biocxml(pmid_list, api_key=pubmed_api_key)\n",
    "#print(res)\n",
    "res = [parse_bioc_xml(r) for r in res]\n",
    "# transform the parsed xml into paper content\n",
    "papers = []\n",
    "for parsed in res:\n",
    "    paper_content = []\n",
    "    for parsed_ in parsed[\"passage\"]:\n",
    "        paper_content.append(parsed_['content'])\n",
    "    paper_content = \"\\n\".join(paper_content)\n",
    "    papers.append(paper_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ac1dabb-884d-4bd7-9ce3-edd08f6c9e35",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1672, 1652, 27171, 2627, 46819]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[len(i) for i in papers]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6a9ab051-ac79-4050-93d9-7008b5074ed6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.',\n",
       " 'Background: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic',\n",
       " 'challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,',\n",
       " 'analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts. Conclusion: Our',\n",
       " 'a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts. Conclusion: Our findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to',\n",
       " 'Our findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splited = split_text_into_chunks(papers[1], chunk_size=600,\n",
    "                                             chunk_overlap=300)\n",
    "splited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ac23d300-fc62-407f-8265-409a50f8a216",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[{'name': 'Crizotinib treatment effectiveness',\n",
       "   'value': 'initial tumor regression followed by disease progression',\n",
       "   'source_id': [2],\n",
       "   'cited_blocks': ['challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,']}],\n",
       " [{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'More sensitive to low-risk group',\n",
       "   'source_id': [5],\n",
       "   'cited_blocks': ['performed in risk groups. Expression of signature genes in CRC cells was examined using quantitative reverse transcription polymerase chain reaction (qRT-PCR). There were 10 CRC-related RRAGs found. Prognostic models were built with RRAGs. Based on immune infiltration analysis, low-risk populations showed significantly greater levels of immune infiltration (P P ARHGEF17 mutation rate was 6%. Medications such as CGP-082996, Dasatinib, Erlotinib, and Salubrinal were more sensitive to high-risk group, whereas drugs such as FTI-277, DMOG, and Crizotinib were more sensitive to low-risk group.']}],\n",
       " [{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'NP',\n",
       "   'source_id': [],\n",
       "   'cited_blocks': []}],\n",
       " [{'name': 'main_results_crizotinib',\n",
       "   'value': 'no objective responses observed in RASMT advanced CRC patients',\n",
       "   'source_id': [7],\n",
       "   'cited_blocks': ['with low baselineÂ\\xa0KRASMT allele frequency. CONCLUSIONS: Combination binimetinib/crizotinib showed a poor tolerability with no objective responses observed in RASMT advanced CRC patients. EudraCT-Number: 2014-000463 -â\\x80\\x8940 (20/06/2014: A Sequential Phase I study of MEK1/2 inhibitors PD- 0325901 or Binimetinib combined with cMET inhibitor Crizotinib in RAS Mutant and RAS Wild Type with aberrant c-MET).']}],\n",
       " [{'name': 'Crizotinib_effectiveness',\n",
       "   'value': 'selectively inhibited the growth of ARID1A-deficient CRC cells',\n",
       "   'source_id': [80, 142],\n",
       "   'cited_blocks': ['validated the synthetic lethal effect of the c-MET inhibitor PHA-665752 (PHA) in two ARID1A knockout (ARID1A-KO) clones, where PHA selectively inhibited the growth of ARID1A-KO HCT116 cells (Fig.Â\\xa01D). Moreover, two FDA-approved c-MET inhibitors, crizotinib, and cabozantinib, also showed synthetic lethality in ARID1A-KO HCT116 clones (Fig.Â\\xa01E, F). These results suggest that c-MET is a potential synthetic lethal target for ARID1A in CRC cells.',\n",
       "    'synthetic lethality in ARID1A-deficient CRC. Notably, FDA-approved c-MET inhibitors, such as crizotinib and cabozantinib, have demonstrated enhanced cytotoxicity in ARID1A-deficient cancers, facilitating the translation of these approved drugs for repurposing in ARID1A-targeted precision cancer medicine.']}]]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "api = StudyCharacteristicsExtraction()\n",
    "extracted = api.run(\n",
    "    papers=papers,\n",
    "    fields=['Results, string, main results of the study regarding Crizotinib, usually about treatment effectiveness'],#api.DEFAULT_FIELDS,\n",
    "    llm='Qwen/Qwen3-32B',#\"openai-gpt-4o\",\n",
    "    chunk_size=600,\n",
    "    chunk_overlap=300,\n",
    "    thinking=False,\n",
    ")\n",
    "extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "dc85b6da-8fe9-4484-8430-e55e239b2525",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'More sensitive to low-risk group',\n",
       "   'source_id': [4],\n",
       "   'cited_blocks': ['performed in risk groups. Expression of signature genes in CRC cells was examined using quantitative reverse transcription polymerase chain reaction (qRT-PCR). There were 10 CRC-related RRAGs found. Prognostic models were built with RRAGs. Based on immune infiltration analysis, low-risk populations showed significantly greater levels of immune infiltration (P P ARHGEF17 mutation rate was 6%. Medications such as CGP-082996, Dasatinib, Erlotinib, and Salubrinal were more sensitive to high-risk group, whereas drugs such as FTI-277, DMOG, and Crizotinib were more sensitive to low-risk group.']}],\n",
       " [{'name': 'Crizotinib treatment effectiveness',\n",
       "   'value': 'initial tumor regression followed by disease progression',\n",
       "   'source_id': [1],\n",
       "   'cited_blocks': ['challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,']}]]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "80601546-e32e-4b65-b837-b5dec0abdfc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm_reasoning = ChatOpenAI(\n",
    "    base_url=BASE_URL,\n",
    "    api_key=HUGGINGFACE_HUB_TOKEN,\n",
    "    model=MODEL_NAME_HF,\n",
    "    streaming=True,\n",
    "    temperature=0,\n",
    "    #http_client=httpx_client\n",
    "    http_async_client=httpx.AsyncClient(verify=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "87efd589-f082-4326-bc29-42ab4cd01c9a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<think>\n",
      "\n",
      "</think>\n",
      "\n",
      "```json\n",
      "[\n",
      "    {\n",
      "        \"name\": \"Crizotinib effectiveness\",\n",
      "        \"value\": \"More sensitive to low-risk group\",\n",
      "        \"source_id\": [3]\n",
      "    }\n",
      "]\n",
      "```\n"
     ]
    }
   ],
   "source": [
    "prompt = ChatPromptTemplate.from_messages([\n",
    "    (\"user\", '/no_think {text}'),\n",
    "    # (\"ai\", \"Good.\")\n",
    "])\n",
    "\n",
    "chain = prompt | llm_reasoning\n",
    "response = await chain.ainvoke({\"text\": ff})\n",
    "\n",
    "#fin_condition = response.content.strip('<think>\\n\\n</think>\\n\\n')\n",
    "print(response.content)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:llms]",
   "language": "python",
   "name": "conda-env-llms-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
