{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "26194509-66fb-4e01-b3c6-47c7952a13c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv, find_dotenv\n",
    "\n",
    "load_dotenv(find_dotenv(usecwd=True))\n",
    "\n",
    "BASE_URL = \"https://gigachat.devices.sberbank.ru/api/v1\"\n",
    "#base_gg = \"https://router.huggingface.co/v1\"\n",
    "#\"http://a6k2.dgx:34000/v1\"\n",
    "\n",
    "BASE_URL = \"https://router.huggingface.co/v1\"\n",
    "\n",
    "#\"qwen3-32b\"\n",
    "# gigachat/GigaChat\n",
    "MODEL_NAME = os.getenv(\"MODEL_NAME\", 'GigaChat')\n",
    "\n",
    "\n",
    "\n",
    "MODEL_NAME_HF ='meta-llama/Llama-3.3-70B-Instruct'\n",
    "MODEL_NAME_HF ='Qwen/Qwen3-32B'\n",
    "\n",
    "MODEL_TEMPERATURE_HF =1\n",
    "\n",
    "os.environ[\"OPENAI_API_KEY\"] = os.getenv(\"GIGACHAT_AUTH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a43a81f3-e83a-4a63-89a6-bd0cc608eff0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MDE5YmM2MjctOWU5ZS03OWE0LWE0OGYtY2FhMTU4MzYxZDgzOjg2MjAwMDgzLThkNzMtNGU0ZS1hNzIxLTM3YWY5ZWY4ZmZlNg=='"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getenv(\"GIGACHAT_AUTH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e14f07d4-cfbd-4129-8f11-5bf5c74d31cf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "from langchain_community.document_loaders import PyMuPDFLoader\n",
    "from pydantic import BaseModel, Field\n",
    "from langchain_core.output_parsers import PydanticOutputParser, StrOutputParser, BaseOutputParser\n",
    "from textwrap import dedent\n",
    "\n",
    "from langchain_openai import ChatOpenAI\n",
    "from langchain_core.messages import HumanMessage\n",
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "import httpx\n",
    "import json\n",
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "from langchain_pymupdf4llm import PyMuPDF4LLMLoader, PyMuPDF4LLMParser\n",
    "import pymupdf4llm\n",
    "\n",
    "from trialmind.pubmed import pmid2papers, PubmedAPIWrapper, pmid2biocxml, parse_bioc_xml\n",
    "from trialmind.api import StudyCharacteristicsExtraction, ScreeningCriteriaGeneration\n",
    "from trialmind.retrievers import split_text_into_chunks\n",
    "\n",
    "import requests\n",
    "import uuid\n",
    "from openai import OpenAI\n",
    "\n",
    "import time\n",
    "import extract\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5301994a-a15c-47ce-b6b8-ce0ceae9e67b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### clinical trials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7eb1ed4e-88ad-454d-a841-868d8d03925b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Colorectal cancer',\n",
       " ['Crizotinib',\n",
       "  'Lorlatinib',\n",
       "  'Necitumumab',\n",
       "  'Nimotuzumab',\n",
       "  'Panitumumab',\n",
       "  'Cetuximab',\n",
       "  'Brigatinib',\n",
       "  'Ramucirumab',\n",
       "  'Pomalidomide',\n",
       "  'Toremifene',\n",
       "  'Denosumab',\n",
       "  'Duvelisib',\n",
       "  'Inavolitib',\n",
       "  'Bevacizumab',\n",
       "  'Anastrozole',\n",
       "  'Exemestane'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"docs/CC_215_L00_DL_fusion_fixed_oncobox_ru.pdf\"\n",
    "fin_condition, treatements_eng = extract.info_from_doc(file_path)\n",
    "fin_condition, treatements_eng"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3c68437-54be-4611-9aa0-d7f6a379a924",
   "metadata": {},
   "source": [
    "1. загрузить док\n",
    "2. взять 1-ую страницу; рег.выражениями взять диагноз\n",
    "-- с ЛЛМ перевести на англ.\n",
    "3. найти страницу с таблицей; или взять таблицы с PyMuPDFLoader\n",
    "4. извлечь таблицу"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "e04c59f8-3cbd-4491-9154-56edbc434e12",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Crizotinib OR Lorlatinib OR Necitumumab OR Nimotuzumab OR Panitumumab OR Cetuximab OR Brigatinib OR Ramucirumab OR Pomalidomide OR Toremifene OR Denosumab OR Duvelisib OR Inavolitib OR Bevacizumab OR Anastrozole OR Exemestane'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "' OR '.join(treatements_eng)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6b530143-823c-495a-8e2a-7039031a23b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0b55cf5c-e883-4a42-ad8f-08633873be8f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_studies = extract.get_clinicaltrials(fin_condition, \n",
    "                                      #' OR '.join(treatements_eng), \n",
    "                                       treatements_eng[0],  \n",
    "                                      max_studies=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30556d38-abf1-4810-a15e-46d3eef966dc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>results</th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>status</th>\n",
       "      <th>conditions</th>\n",
       "      <th>study_type</th>\n",
       "      <th>phases</th>\n",
       "      <th>interventions</th>\n",
       "      <th>outcomes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT02465060</td>\n",
       "      <td>Targeted Therapy Directed by Genetic Testing i...</td>\n",
       "      <td>ACTIVE_NOT_RECRUITING</td>\n",
       "      <td>[Advanced Lymphoma, Advanced Malignant Solid N...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[adavosertib, afatinib, afatinib dimaleate, bi...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT03947385</td>\n",
       "      <td>Study of IDE196 in Patients With Solid Tumors ...</td>\n",
       "      <td>RECRUITING</td>\n",
       "      <td>[Metastatic Uveal Melanoma, Cutaneous Melanoma...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE1, PHASE2]</td>\n",
       "      <td>[ide196, binimetinib, crizotinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT02568267</td>\n",
       "      <td>Basket Study of Entrectinib (RXDX-101) for the...</td>\n",
       "      <td>ACTIVE_NOT_RECRUITING</td>\n",
       "      <td>[Breast Cancer, Cholangiocarcinoma, Colorectal...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[entrectinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT06214793</td>\n",
       "      <td>Taletrectinib in Previously Treated Metastatic...</td>\n",
       "      <td>SUSPENDED</td>\n",
       "      <td>[Breast Cancer, Metastatic Breast Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[taletrectinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT03792568</td>\n",
       "      <td>ALK Inhibitor in Metastatic Colorectal Cancer ...</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>[Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT05725200</td>\n",
       "      <td>Study to Investigate Outcome of Individualized...</td>\n",
       "      <td>RECRUITING</td>\n",
       "      <td>[Metastatic Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[alectinib, cetuximab, crizotinib, dasatinib, ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>NCT02510001</td>\n",
       "      <td>MEK and MET Inhibition in Colorectal Cancer</td>\n",
       "      <td>COMPLETED</td>\n",
       "      <td>[Solid Tumor, Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE1]</td>\n",
       "      <td>[pf-02341066, pd-0325901, binimetinib]</td>\n",
       "      <td>[{'type': 'PRIMARY', 'title': 'Maximal Tolerat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   results           id                                              title  \\\n",
       "0    False  NCT02465060  Targeted Therapy Directed by Genetic Testing i...   \n",
       "1    False  NCT03947385  Study of IDE196 in Patients With Solid Tumors ...   \n",
       "2    False  NCT02568267  Basket Study of Entrectinib (RXDX-101) for the...   \n",
       "3    False  NCT06214793  Taletrectinib in Previously Treated Metastatic...   \n",
       "4    False  NCT03792568  ALK Inhibitor in Metastatic Colorectal Cancer ...   \n",
       "5    False  NCT05725200  Study to Investigate Outcome of Individualized...   \n",
       "6     True  NCT02510001        MEK and MET Inhibition in Colorectal Cancer   \n",
       "\n",
       "                  status                                         conditions  \\\n",
       "0  ACTIVE_NOT_RECRUITING  [Advanced Lymphoma, Advanced Malignant Solid N...   \n",
       "1             RECRUITING  [Metastatic Uveal Melanoma, Cutaneous Melanoma...   \n",
       "2  ACTIVE_NOT_RECRUITING  [Breast Cancer, Cholangiocarcinoma, Colorectal...   \n",
       "3              SUSPENDED          [Breast Cancer, Metastatic Breast Cancer]   \n",
       "4                UNKNOWN                                [Colorectal Cancer]   \n",
       "5             RECRUITING                     [Metastatic Colorectal Cancer]   \n",
       "6              COMPLETED                   [Solid Tumor, Colorectal Cancer]   \n",
       "\n",
       "       study_type            phases  \\\n",
       "0  INTERVENTIONAL          [PHASE2]   \n",
       "1  INTERVENTIONAL  [PHASE1, PHASE2]   \n",
       "2  INTERVENTIONAL          [PHASE2]   \n",
       "3  INTERVENTIONAL          [PHASE2]   \n",
       "4  INTERVENTIONAL              [NA]   \n",
       "5  INTERVENTIONAL          [PHASE2]   \n",
       "6  INTERVENTIONAL          [PHASE1]   \n",
       "\n",
       "                                       interventions  \\\n",
       "0  [adavosertib, afatinib, afatinib dimaleate, bi...   \n",
       "1                  [ide196, binimetinib, crizotinib]   \n",
       "2                                      [entrectinib]   \n",
       "3                                    [taletrectinib]   \n",
       "4                                                 []   \n",
       "5  [alectinib, cetuximab, crizotinib, dasatinib, ...   \n",
       "6             [pf-02341066, pd-0325901, binimetinib]   \n",
       "\n",
       "                                            outcomes  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2                                                NaN  \n",
       "3                                                NaN  \n",
       "4                                                NaN  \n",
       "5                                                NaN  \n",
       "6  [{'type': 'PRIMARY', 'title': 'Maximal Tolerat...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_studies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37f1ccfb-f682-457e-bf61-1bed5aad563c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### outcome results\n",
    "Взяли все результаты, отдельно сохранили испытания с результатами, и з них отбираем рез-ты только с unit==participants\n",
    "\n",
    "filter by interventional study type?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "c5e73eef-7cac-4115-93c7-34ba0ee46532",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'type': 'PRIMARY',\n",
       "  'title': 'Clinical Response to Binimetinib Combined With PF-02341066',\n",
       "  'description': 'To investigate response to treatment with RPII dose of Binimetinib with Crizotinib (PF-02341066), in patients with a) RASMT CRC or b) RASWT/cMET mut amplified CRC or c) RASWT/c-MET over-expressed CRC, as defined by stable, partially or completely responding disease, per Response Evaluation Criteria In Solid Tumors Criteria (RECIST v1.1) for target lesions and assessed by CT: Complete Response (CR), Disappearance of all target lesions; Partial Response (PR), \\\\>=30% decrease in the sum of the longest diameter of target lesions; Stable Disease (SD), Neither sufficient shrinkage to qualify for PR nor sufficient increase (\\\\>=20%) to qualify for Progressive Disease; Overall Response (OR) = CR + PR + SD',\n",
       "  'populationDescription': 'Evaluable patients for the primary outcome are those patients who complete a response assessment after cycle 1 of treatment, or who progress early on treatment. 30 of the 36 recruited patients had a response assessment after cycle 1 or progressed early on treatment, and hence these 30 patients are evaluable for the primary analysis.',\n",
       "  'reportingStatus': 'POSTED',\n",
       "  'paramType': 'COUNT_OF_PARTICIPANTS',\n",
       "  'unitOfMeasure': 'Participants',\n",
       "  'timeFrame': 'Dose Expansion phase: change from baseline and up to 12 months.',\n",
       "  'groups': [{'id': 'OG000',\n",
       "    'title': 'Dose Expansion Phase',\n",
       "    'description': 'Binimetinib 30mg BD interval dose administration Days 1-21 every 28 days PF-02341066 (Crizotinib) 250mg OD Days 1-28 continuously Dosage determined following the recommended Phase II dose identification in the dose escalation phase.\\n\\nPF-02341066: PF-02341066 (Crizotinib) 250mg OD or 200mg BD or 250mg BD Days 1-28 continuously\\n\\nBinimetinib: Binimetinib 30mg or 45 mg BD either continuously or Days 1-21 every 28 days'}],\n",
       "  'denoms': [{'units': 'Participants',\n",
       "    'counts': [{'groupId': 'OG000', 'value': '30'}]}],\n",
       "  'classes': [{'categories': [{'title': 'Stable Disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '7'}]},\n",
       "     {'title': 'Progressive Disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '22'}]},\n",
       "     {'title': 'Early death from malignant disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '1'}]}]}]}]"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outcome_participants = [obj for item_main in results_studies for obj in item_main['resultsSection']['outcomeMeasuresModule'\n",
    "                                           ]['outcomeMeasures'\n",
    "                                            ] if(obj.get('unitOfMeasure','') == 'Participants')] \n",
    "outcome_participants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9303a48c-9dd0-418b-8c9f-3708b4f9148a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "db6360d3-76fa-4607-8764-0a705af4464c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<think>\n",
      "\n",
      "</think>\n",
      "\n",
      "Here is a concise summary of the efficiency statistics for the clinical response to Binimetinib combined with PF-02341066 (Crizotinib) in the Dose Expansion Phase:\n",
      "\n",
      "- **Total Evaluable Patients**: 30 participants.\n",
      "- **Stable Disease (SD)**: 7 patients (23.3%).\n",
      "- **Progressive Disease (PD)**: 22 patients (73.3%).\n",
      "- **Early Death from Malignant Disease**: 1 patient (3.3%).\n",
      "\n",
      "These outcomes were assessed using RECIST v1.1 criteria after cycle 1 of treatment.\n"
     ]
    }
   ],
   "source": [
    "prompt2 = ChatPromptTemplate.from_messages([\n",
    "    (\"system\", \"You are a helpful medical assistant. Extract information about the efficiency statistics from each outcome measure. Give a concise description with percentages /no_think\"),\n",
    "    (\"human\", \"{text}\"),\n",
    "    # (\"ai\", \"Good.\")\n",
    "])\n",
    "\n",
    "chain2 = prompt2 | llm_reasoning\n",
    "response_res = await chain2.ainvoke({\"text\": outcome_participants[0]})\n",
    "print(response_res.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84f5d879-2920-4ec3-bac4-5ae64fb55099",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e706ee64-c65f-40a9-8606-8e71c97fa5d6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### results in a given format?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f1d0545-8778-43fb-9af8-da1730e79453",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class Info(BaseModel):\n",
    "    diagnose: str = Field(..., description=\"База клинических испытаний\")\n",
    "    treatments: str = Field(..., description=\"Высшие позиции рейтинга таргетных препаратов\")\n",
    "\n",
    "parser = PydanticOutputParser(pydantic_object=Info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d5b0edc-bb8f-4fcd-b443-36745ad7eaf7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "prompt = ChatPromptTemplate.from_messages([\n",
    "    (\n",
    "        \"system\", \n",
    "        dedent(\n",
    "            \"\"\"\n",
    "            You are a helpful assistant. \n",
    "            Your goal is to extract related information from the text provided by the user.\n",
    "            Return the information in the following JSON-format.\n",
    "            {fmt}\n",
    "            \"\"\"\n",
    "        )\n",
    "    ),\n",
    "    (\n",
    "        \"human\", \n",
    "        dedent(\n",
    "            \"\"\"\n",
    "            Text: {text}\"\n",
    "            \"\"\"\n",
    "        )\n",
    "    ),\n",
    "])\n",
    "chain = (\n",
    "    prompt \n",
    "    | llm_reasoning \n",
    "    | parser\n",
    ")\n",
    "\n",
    "\n",
    "response = await chain.ainvoke({\"text\": text, \"fmt\": parser.get_format_instructions()})\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d963b7db-0c01-4b5f-af74-d7d566ae8a54",
   "metadata": {
    "tags": []
   },
   "source": [
    "### papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1e430e6-c0a6-4a2f-9e92-2b0ca80bdd6e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fin_condition = 'Colorectal cancer'\n",
    "treatements_eng =  ['Crizotinib','Lorlatinib',]\n",
    "pubmed_api_key= '6892c4129cef143ff92d11533848d2e0d908'#os.getenv(\"PUBMED_API_KEY\", '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ceab5c26-765a-420c-82a6-8bfbacdb36f6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['41268440'],\n",
       " 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=(Colorectal cancer)+AND+(Crizotinib)&retmax=1&retmode=json&api_key=6892c4129cef143ff92d11533848d2e0d908',\n",
       " '67')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "search_api = PubmedAPIWrapper()\n",
    "# page_size is the max number of records to return!!!! not pages!\n",
    "tmp_inputs = {\n",
    "        \"page_size\": 1,\n",
    "        \"keyword_map\": {'conditions':[fin_condition], \n",
    "                        'treatments':[treatements_eng[0]]\n",
    "                       },\n",
    "        \"keywords\": {\n",
    "            \"OPERATOR\": 'AND'\n",
    "        }\n",
    "}\n",
    "\n",
    "time.sleep(.5)\n",
    "response = search_api.build_search_query_and_get_pmid(tmp_inputs, \n",
    "                                                      api_key=pubmed_api_key)\n",
    "response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a54a4f2c-d723-4a4a-a1b1-77eb50eeaea8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Year</th>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th>Title</th>\n",
       "      <th>Publication Type</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>41268440</td>\n",
       "      <td>Frontiers in pharmacology</td>\n",
       "      <td>2025</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Case Report: HGF and NF1 mutations as putative...</td>\n",
       "      <td>Case Reports, Journal Article</td>\n",
       "      <td>Zhitao Chen, Shan Luo, Yangjun Gu, Qiyong Li</td>\n",
       "      <td>Hepatocellular carcinoma (HCC) is a highly agg...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                    Journal  Year Month Day  \\\n",
       "0  41268440  Frontiers in pharmacology  2025             \n",
       "\n",
       "                                               Title  \\\n",
       "0  Case Report: HGF and NF1 mutations as putative...   \n",
       "\n",
       "                Publication Type  \\\n",
       "0  Case Reports, Journal Article   \n",
       "\n",
       "                                        Authors  \\\n",
       "0  Zhitao Chen, Shan Luo, Yangjun Gu, Qiyong Li   \n",
       "\n",
       "                                            Abstract  \n",
       "0  Hepatocellular carcinoma (HCC) is a highly agg...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_papers = pmid2papers(pmid_list=['41268440'],#response[0], \n",
    "                        api_key=pubmed_api_key)\n",
    "df_papers[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dd4f5664-3983-4bbb-af9f-7f460b87e5e3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Colorectal cancer', ['Crizotinib', 'Lorlatinib'])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_condition, treatements_eng"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f15bedd0-7b97-4696-ac71-2b9ddc4ef0d6",
   "metadata": {},
   "source": [
    "### extract results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "365da54d-32aa-445c-b27b-0b6b2ae80cf6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fin_condition = 'Colorectal cancer'\n",
    "treatements_eng =  ['Crizotinib','Lorlatinib',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "033c3e2a-28c7-4d05-b135-1078f19e8e3d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MDE5YmM2MjctOWU5ZS03OWE0LWE0OGYtY2FhMTU4MzYxZDgzOjg2MjAwMDgzLThkNzMtNGU0ZS1hNzIxLTM3YWY5ZWY4ZmZlNg=='"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getenv(\"GIGACHAT_AUTH\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f9825e3-9cba-4bc3-9278-f9d294302f69",
   "metadata": {},
   "source": [
    "### get sber api"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "66ffb1de-672b-4c10-acfc-bcade11f8c7d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\urllib3\\connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host 'ngw.devices.sberbank.ru'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'eyJjdHkiOiJqd3QiLCJlbmMiOiJBMjU2Q0JDLUhTNTEyIiwiYWxnIjoiUlNBLU9BRVAtMjU2In0.Z_GQStOEb6STz0PuIvi2c4hZzSijwZMB8XO6ONX3M3Kp2KAbSZJrOU76w4XZxfoD_VXFQVMwBKUKFii1fj2WPPbang57m8icdbqkMRUBEL71dyAHqIg1il6dXoAbKu4iI1Ps9guhPNrwU26pQ64eCJA4agcH2SL3qeMruYPz7lVOt7YrQS4O9cCUO7YnKDLa-BCffEEv_B_H-oGryLeJ_LnvPChX49dLXGyBRdFF9DQ6qI3eyZ9KsU4aSRYpeYiozapNPg0Isdi9_dQoX0jvluGS_SurQ-gmyiY7I3_DSgYHrIDqeGa8C0AsiROO7eHuYxOzOXUkPowp8pRQAMa7Yw.MWSJ9br9Ka7Lay1gHVCA2A.wH5v2iXQ3Qgk0Jzkz_S2UrjKE6YxvE37CDYnzqsYFI_G5Oh8qgun3_W1YJxACkqsx787825R30XwqMrDVYON12t8hYQJiw9uPYI8NmS3IOOuK5rKeILa95NdKxAfZmB5BJHfnDNJPLiTvNPATu7YldXQdHCBOd7sx1kyreqCuNJAuidGKEI9qB7vIBesN9nmA83ESQ2QiNMm-QKnqCu0LVm1iJAZ9BWDOEZuXkHvcSpeglp_M6fpWm5CkrdKZkPpExZUtX7Yg8qZ-ErNIem4wfSaNemAhAR9YlieUgwKyiKWPMMblc80iMq8MhDCH5OiBTzxtVJyVSmrRLkc-Iw5_jC7qSMuhiyLXHdFFO-Dn6xENp5pgB7wng5JpqE0eZw2wyBZ5LX1IkbGk_j6XG-REB-G6nyvYDzYYYPL8F7aYalov3BANykGScL57AG7VfXKgKKt5giG1LdBay3ImQKombw-byieqw1ttRw-jHRhlyAFncbmLYlS7UCaBvFLeW_XWzsG_Hnmv4ZBCo712njTbG_6yk61n-gTRIJTgZLCL0PHFcYjvORig_GCoaLC9ys1QKqVSUZObYFP-naBxMqkP6A0QjXyXg-VLL-IkHLPCHZWwskS21DNTHSQ8pOBaeUbIyCWJiVVmUZpCZpICh1e07dpwCeuDAmEnRVtP-kZo5xRlE5EZearWpud8wmv5CH46SpCvfHXd2UxEfveiR1rF2AO7UtKmbz8aQ4r66sFYkg.S2M8gSDzobYP5H9up3UpU6TeIMNUxnpyS6s7zRFA4VA'"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "url = \"https://ngw.devices.sberbank.ru:9443/api/v2/oauth\"\n",
    "# Создадим идентификатор UUID (36 знаков)\n",
    "rq_uid = str(uuid.uuid4())\n",
    "payload={\n",
    "  'scope': 'GIGACHAT_API_PERS'\n",
    "}\n",
    "headers = {\n",
    "  'Content-Type': 'application/x-www-form-urlencoded',\n",
    "  'Accept': 'application/json',\n",
    "  'RqUID': rq_uid,\n",
    "  'Authorization': f'Basic {os.getenv(\"GIGACHAT_AUTH\")}'\n",
    "}\n",
    "\n",
    "response = requests.request(\"POST\", url, headers=headers, data=payload, verify=False)\n",
    "os.environ[\"OPENAI_API_KEY\"] = response.json()['access_token']\n",
    "\n",
    "os.getenv(\"OPENAI_API_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "720ecef2-268b-4b8e-8241-96b44d26d0bf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4e632a26-387b-4c21-b67c-5408ce327484",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "os.environ[\"BASE_URL\"] = \"https://gigachat.devices.sberbank.ru/api/v1\"\n",
    "os.environ[\"MODEL_NAME\"] ='GigaChat'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "edf59fa4-0dbb-4fd6-8af9-dbaa5190e776",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://gigachat.devices.sberbank.ru/api/v1'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getenv(\"BASE_URL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f7987e16-3579-4cf1-b9a0-ef66a903aeb9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "call_openai https://gigachat.devices.sberbank.ru/api/v1\n",
      "api_call_single https://gigachat.devices.sberbank.ru/api/v1/\n",
      "```json\n",
      "{\n",
      "  \"ELIGIBILITY_ANALYSIS\": [\n",
      "    \"Patients must have colorectal cancer — Studies focusing on other types of cancers would not provide relevant data.\",\n",
      "    \"Studies involving patients receiving crizotinib as an intervention — Other treatments or interventions would not contribute to the analysis.\",\n",
      "    \"Comparative studies where crizotinib is compared against another treatment or placebo — Studies without comparison groups cannot assess effectiveness relative to alternatives.\",\n",
      "    \"Studies reporting outcomes related to patient survival, response rates, toxicity, or quality of life — Only studies measuring relevant outcomes can inform decision-making.\",\n",
      "    \"Published studies conducted after 2010 — Older studies may lack current evidence or methodology.\"\n",
      "  ],\n",
      "  \"TITLE_CRITERIA\": [\n",
      "    \"Does the title mention 'colorectal cancer'?\",\n",
      "    \"Does the title include 'crizotinib' or similar terms like 'Crizolta', 'PF-02341066'?\",\n",
      "    \"Is there any indication in the title that the study compares crizotinib to another treatment?\"\n",
      "  ],\n",
      "  \"CONTENT_CRITERIA\": [\n",
      "    \"Are the methods section and results clearly described in the paper?\",\n",
      "    \"Do the results specifically report outcomes such as overall survival, progression-free survival, response rate, or adverse events?\",\n",
      "    \"Was the study randomized controlled trial (RCT), prospective cohort study, or retrospective case-control study? — Preferably RCTs for higher reliability.\"\n",
      "  ]\n",
      "}\n",
      "```  \n",
      "\n",
      "This structured approach ensures that only relevant studies are selected for inclusion in the meta-analysis, ensuring high-quality and reliable findings.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'title_criteria': [\"Does the title mention 'colorectal cancer'?\",\n",
       "  \"Does the title include 'crizotinib' or similar terms like 'Crizolta', 'PF-02341066'?\",\n",
       "  'Is there any indication in the title that the study compares crizotinib to another treatment?'],\n",
       " 'content_criteria': ['Are the methods section and results clearly described in the paper?',\n",
       "  'Do the results specifically report outcomes such as overall survival, progression-free survival, response rate, or adverse events?',\n",
       "  'Was the study randomized controlled trial (RCT), prospective cohort study, or retrospective case-control study? — Preferably RCTs for higher reliability.'],\n",
       " 'eligibility_analysis': ['Patients must have colorectal cancer — Studies focusing on other types of cancers would not provide relevant data.',\n",
       "  'Studies involving patients receiving crizotinib as an intervention — Other treatments or interventions would not contribute to the analysis.',\n",
       "  'Comparative studies where crizotinib is compared against another treatment or placebo — Studies without comparison groups cannot assess effectiveness relative to alternatives.',\n",
       "  'Studies reporting outcomes related to patient survival, response rates, toxicity, or quality of life — Only studies measuring relevant outcomes can inform decision-making.',\n",
       "  'Published studies conducted after 2010 — Older studies may lack current evidence or methodology.']}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test the apis\n",
    "\n",
    "api = ScreeningCriteriaGeneration()\n",
    "ecs = api.run(\n",
    "    population = f\"Patients with {fin_condition} undergoing treatment with {treatements_eng[0]}\",\n",
    "    intervention = f\"{treatements_eng[0]}\",\n",
    "    comparator = \"\",\n",
    "    outcome = \"\",\n",
    "    llm = os.getenv(\"MODEL_NAME\"),\n",
    "    num_title_criteria=3,\n",
    "    num_abstract_criteria=3,\n",
    "    thinking=False,\n",
    ")\n",
    "ecs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f2858acb-a1e3-4d89-b8af-f1b7ce0cf1d4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC.\\nWe present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance.\\nNGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts.\\nOur findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from trialmind.api import LiteratureScreening\n",
    "api = LiteratureScreening()\n",
    "\n",
    "papers = df_papers[0].iloc[:3][\"Title\"] + \": \" + df_papers[0].iloc[:3][\"Abstract\"].fillna(\"\") # important to fillna\n",
    "papers = papers.tolist()\n",
    "papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d94360-c951-4e9a-b37f-8fc6e2eea203",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e604d682-87c0-4bcb-97c0-0c4109f1bdd3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[\"Does the title mention 'colorectal cancer'?\",\n",
       " \"Does the title include 'crizotinib' or similar terms like 'Crizolta', 'PF-02341066'?\",\n",
       " 'Is there any indication in the title that the study compares crizotinib to another treatment?',\n",
       " 'Are the methods section and results clearly described in the paper?',\n",
       " 'Do the results specifically report outcomes such as overall survival, progression-free survival, response rate, or adverse events?',\n",
       " 'Was the study randomized controlled trial (RCT), prospective cohort study, or retrospective case-control study? — Preferably RCTs for higher reliability.']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "title_criteria = [\"Does the title mention 'colorectal cancer'?\",\n",
    "\"Does the title mention 'crizotinib'?\",\n",
    "\"Is there any indication in the title that the study compares crizotinib to another treatment?\"\n",
    "]\n",
    "\n",
    "content_criteria = [\"Are the methods section and results clearly described in the paper?\",\n",
    "\"Does the study specify the type of colorectal cancer (e.g., adenocarcinoma)?\",\n",
    "\"Were the outcomes measured over a sufficient duration to assess long-term effects?\"\n",
    "  ]\n",
    "'''\n",
    "title_criteria = ecs['title_criteria']\n",
    "content_criteria = ecs['content_criteria']\n",
    "title_criteria+content_criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54842268-da25-4f52-b604-fd1b0473a2e9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6e510df2-8174-4105-bcaa-00767e4560ba",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "api_function_call_single https://gigachat.devices.sberbank.ru/api/v1/\n",
      "[[{'role': 'user', 'content': '/no_think \\n# CONTEXT #\\nYou are a clinical specialist tasked with assessing research papers for inclusion in a meta-analysis based on specific eligibility criteria.\\n\\n# OBJECTIVE #\\nEvaluate each criterion of a given paper to determine its eligibility for inclusion in the meta-analysis. Provide a list of decisions (\"YES\", \"NO\", or \"UNCERTAIN\") for each eligibility criterion. You must deliver exactly 6 responses.\\n\\n# IMPORTANT NOTE #\\nIf the information within the provided paper content is insufficient to conclusively evaluate a criterion, you must opt for \"UNCERTAIN\" as your response. Avoid making assumptions or extrapolating beyond the provided data, as accurate and reliable responses are crucial, and fabricating information (hallucinations) could lead to serious errors in the meta-analysis.\\n\\n# PICO FRAMEWORK #\\n- P (Patient, Problem or Population): Patients with Colorectal cancer undergoing treatment with Crizotinib\\n- I (Intervention): Crizotinib\\n- C (Comparison): \\n- O (Outcome): \\n\\n# PAPER DETAILS #\\n- Provided Paper: Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC.\\nWe present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance.\\nNGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts.\\nOur findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.\\n\\n# EVALUATION CRITERIA #\\n- Number of Criteria: 6\\n- Criteria for Inclusion: [\"1. Does the title mention \\'colorectal cancer\\'?\", \"2. Does the title include \\'crizotinib\\' or similar terms like \\'Crizolta\\', \\'PF-02341066\\'?\", \\'3. Is there any indication in the title that the study compares crizotinib to another treatment?\\', \\'4. Are the methods section and results clearly described in the paper?\\', \\'5. Do the results specifically report outcomes such as overall survival, progression-free survival, response rate, or adverse events?\\', \\'6. Was the study randomized controlled trial (RCT), prospective cohort study, or retrospective case-control study? — Preferably RCTs for higher reliability.\\']\\n\\n# RESPONSE FORMAT #\\nReturn the information in a JSON object containing a list of decisions for each of the 6 eligibility criteria. Each decision should directly correspond to one of the criteria and be listed in the order they are presented. Ensure to use \"UNCERTAIN\" wherever the paper does not explicitly support a \"YES\" or \"NO\" decision.\\n\\nFor example:\\n```json\\n{\\n    \"evaluations\": [\"YES\", \"NO\", \"UNCERTAIN\", \"YES\", \"YES\", ...]\\n}\\n```\\n'}]]\n",
      "batch_function_call_openai in asynch [ChatCompletion(id=None, choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='```json\\n{\\n    \"evaluations\": [\\n        \"NO”,   // The title does not mention colorectal cancer\\n        \"YES\",  // Title includes \"crizotinib\"\\n        \"NO\",   // No comparison mentioned in the title\\n        \"YES\",  // Methods and results sections are clearly described\\n        \"NO\",   // Results do not specifically report outcomes such as OS, PFS, RR, or AEs\\n        \"NO”   // Study design is not a randomized controlled trial (RCT)\\n    ]\\n}\\n```', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None))], created=1770371062, model='GigaChat:2.0.28.2', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=116, prompt_tokens=828, total_tokens=944, completion_tokens_details=None, prompt_tokens_details=None, precached_prompt_tokens=2))]\n",
      "response_message ChatCompletionMessage(content='```json\\n{\\n    \"evaluations\": [\\n        \"NO”,   // The title does not mention colorectal cancer\\n        \"YES\",  // Title includes \"crizotinib\"\\n        \"NO\",   // No comparison mentioned in the title\\n        \"YES\",  // Methods and results sections are clearly described\\n        \"NO\",   // Results do not specifically report outcomes such as OS, PFS, RR, or AEs\\n        \"NO”   // Study design is not a randomized controlled trial (RCT)\\n    ]\\n}\\n```', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None)\n",
      "parsed_results in asynch [{}]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[['UNCERTAIN',\n",
       "  'UNCERTAIN',\n",
       "  'UNCERTAIN',\n",
       "  'UNCERTAIN',\n",
       "  'UNCERTAIN',\n",
       "  'UNCERTAIN']]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ec_pred = api.run(\n",
    "    population = f\"Patients with {fin_condition} undergoing treatment with {treatements_eng[0]}\",\n",
    "    intervention = f\"{treatements_eng[0]}\",\n",
    "    comparator = \"\",\n",
    "    outcome = \"\",\n",
    "    llm = os.getenv(\"MODEL_NAME\"),\n",
    "    criteria = title_criteria + content_criteria,\n",
    "    papers = papers[:1], # make for the top-100 for demo\n",
    ")\n",
    "ec_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f6c9066a-dfbd-49fc-84de-1fa91a894c1f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from openai import OpenAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "4b3862e0-f552-42da-83eb-67220629ec5b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "llm_reasoning = OpenAI(\n",
    "    base_url=os.getenv(\"BASE_URL\"),\n",
    "    api_key=os.getenv(\"OPENAI_API_KEY\"),\n",
    "    http_client=httpx.Client(http2=True, verify=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "d88d6abe-7167-48aa-bfa7-47550946d97b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from trialmind.prompts.screening import LITERATURE_SCREENING_FC\n",
    "from trialmind.llm import _batch_inputs_to_messages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "5b65224b-4cf3-4995-9bad-9875cc58dd1c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The output should be formatted as a JSON instance that conforms to the JSON schema below.\n",
      "\n",
      "As an example, for the schema {\"properties\": {\"foo\": {\"title\": \"Foo\", \"description\": \"a list of strings\", \"type\": \"array\", \"items\": {\"type\": \"string\"}}}, \"required\": [\"foo\"]}\n",
      "the object {\"foo\": [\"bar\", \"baz\"]} is a well-formatted instance of the schema. The object {\"properties\": {\"foo\": [\"bar\", \"baz\"]}} is not well-formatted.\n",
      "\n",
      "Here is the output schema:\n",
      "```\n",
      "{\"$defs\": {\"Criteria\": {\"properties\": {\"decision\": {\"anyOf\": [{\"type\": \"string\"}, {\"enum\": [\"YES\", \"NO\", \"UNCERTAIN\"], \"type\": \"string\"}], \"title\": \"Decision\"}, \"rationale\": {\"description\": \"a rationale for each decision\", \"title\": \"Rationale\", \"type\": \"string\"}}, \"required\": [\"decision\", \"rationale\"], \"title\": \"Criteria\", \"type\": \"object\"}}, \"properties\": {\"criterias\": {\"items\": {\"$ref\": \"#/$defs/Criteria\"}, \"title\": \"Criterias\", \"type\": \"array\"}}, \"required\": [\"criterias\"]}\n",
      "```\n"
     ]
    }
   ],
   "source": [
    "#\"fmt\": \n",
    "from typing import Union, Literal\n",
    "class Criteria(BaseModel):\n",
    "    decision: Union[str, Literal[\"YES\", \"NO\", \"UNCERTAIN\"]]\n",
    "    rationale: str = Field(..., description=\"a rationale for each decision\")\n",
    "\n",
    "class Criterias(BaseModel):\n",
    "    criterias: list[Criteria]\n",
    "\n",
    "parser = PydanticOutputParser(pydantic_object=Criterias)\n",
    "print(parser.get_format_instructions())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a2a1efd-55ca-4a8e-a3cc-4356a091dfe9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "260c8655-11a1-46cb-a578-80de8febb946",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'P': ('Patients with Colorectal cancer undergoing treatment with Crizotinib',),\n",
       "  'I': ('Crizotinib',),\n",
       "  'C': ('',),\n",
       "  'O': ('',),\n",
       "  'paper_content': (((((((['Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC.\\nWe present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance.\\nNGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts.\\nOur findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.'],),),),),),),),\n",
       "  'criteria_text': [\"Does the title mention 'colorectal cancer'?\",\n",
       "   \"Does the title include 'crizotinib' or similar terms like 'Crizolta', 'PF-02341066'?\",\n",
       "   'Is there any indication in the title that the study compares crizotinib to another treatment?',\n",
       "   'Are the methods section and results clearly described in the paper?',\n",
       "   'Do the results specifically report outcomes such as overall survival, progression-free survival, response rate, or adverse events?',\n",
       "   'Was the study randomized controlled trial (RCT), prospective cohort study, or retrospective case-control study? — Preferably RCTs for higher reliability.'],\n",
       "  'num_criteria': 6,\n",
       "  'fin_format': 'The output should be formatted as a JSON instance that conforms to the JSON schema below.\\n\\nAs an example, for the schema {\"properties\": {\"foo\": {\"title\": \"Foo\", \"description\": \"a list of strings\", \"type\": \"array\", \"items\": {\"type\": \"string\"}}}, \"required\": [\"foo\"]}\\nthe object {\"foo\": [\"bar\", \"baz\"]} is a well-formatted instance of the schema. The object {\"properties\": {\"foo\": [\"bar\", \"baz\"]}} is not well-formatted.\\n\\nHere is the output schema:\\n```\\n{\"$defs\": {\"Criteria\": {\"properties\": {\"decision\": {\"anyOf\": [{\"type\": \"string\"}, {\"enum\": [\"YES\", \"NO\", \"UNCERTAIN\"], \"type\": \"string\"}], \"title\": \"Decision\"}, \"rationale\": {\"description\": \"a rationale for each decision\", \"title\": \"Rationale\", \"type\": \"string\"}}, \"required\": [\"decision\", \"rationale\"], \"title\": \"Criteria\", \"type\": \"object\"}}, \"properties\": {\"criterias\": {\"items\": {\"$ref\": \"#/$defs/Criteria\"}, \"title\": \"Criterias\", \"type\": \"array\"}}, \"required\": [\"criterias\"]}\\n```'}]"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "population = f\"Patients with {fin_condition} undergoing treatment with {treatements_eng[0]}\",\n",
    "intervention = f\"{treatements_eng[0]}\",\n",
    "comparator = \"\",\n",
    "outcome = \"\",\n",
    "llm = os.getenv(\"MODEL_NAME\"),\n",
    "criteria = title_criteria + content_criteria,\n",
    "papers = papers[:1], # make for the top-100 for demo\n",
    "batch_inputs = []\n",
    "batch_inputs.append({\n",
    "            \"P\": population,\n",
    "            \"I\": intervention,\n",
    "            \"C\": comparator,\n",
    "            \"O\": outcome,\n",
    "            \"paper_content\": papers[0],\n",
    "            \"criteria_text\": title_criteria + content_criteria,\n",
    "            \"num_criteria\": 6,\n",
    "            \"fin_format\":parser.get_format_instructions(),\n",
    "    })\n",
    "batch_inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "7dde930e-b2a9-4863-bb02-4df944c2e925",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChatPromptTemplate(input_variables=['C', 'I', 'O', 'P', 'criteria_text', 'fin_format', 'num_criteria', 'paper_content'], input_types={}, partial_variables={}, messages=[SystemMessagePromptTemplate(prompt=PromptTemplate(input_variables=['C', 'I', 'O', 'P', 'criteria_text', 'fin_format', 'num_criteria', 'paper_content'], input_types={}, partial_variables={}, template='\\n# CONTEXT #\\nYou are a clinical specialist tasked with assessing research papers for inclusion in a meta-analysis based on specific eligibility criteria.\\n\\n# OBJECTIVE #\\nEvaluate each criterion of a given paper to determine its eligibility for inclusion in the meta-analysis. Provide a list of decisions (\"YES\", \"NO\", or \"UNCERTAIN\") for each eligibility criterion. You must deliver exactly {num_criteria} responses.\\n\\n# IMPORTANT NOTE #\\nIf the information within the provided paper content is insufficient to conclusively evaluate a criterion, you must opt for \"UNCERTAIN\" as your response. Avoid making assumptions or extrapolating beyond the provided data, as accurate and reliable responses are crucial, and fabricating information (hallucinations) could lead to serious errors in the meta-analysis.\\n\\n# PICO FRAMEWORK #\\n- P (Patient, Problem or Population): {P}\\n- I (Intervention): {I}\\n- C (Comparison): {C}\\n- O (Outcome): {O}\\n\\n# PAPER DETAILS #\\n- Provided Paper: {paper_content}\\n\\n# EVALUATION CRITERIA #\\n- Number of Criteria: {num_criteria}\\n- Criteria for Inclusion: {criteria_text}\\n\\n# RESPONSE FORMAT #\\nReturn the information in a JSON object containing a list of decisions for each of the {num_criteria} eligibility criteria. Each decision should directly correspond to one of the criteria and be listed in the order they are presented. Ensure to use \"UNCERTAIN\" wherever the paper does not explicitly support a \"YES\" or \"NO\" decision.\\n\\nThe format:\\n{fin_format}\\n'), additional_kwargs={}), HumanMessagePromptTemplate(prompt=PromptTemplate(input_variables=[], input_types={}, partial_variables={}, template='tell me'), additional_kwargs={})])"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "prompt = ChatPromptTemplate.from_messages([('system',LITERATURE_SCREENING_FC),\n",
    "                                          ('human', 'tell me')])\n",
    "prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "5f78fe4b-1bac-49f3-9f83-6944932e4a75",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "llm_reasoning = ChatOpenAI(\n",
    "    #api_base=os.getenv(\"BASE_URL\"),\n",
    "    base_url=os.getenv(\"BASE_URL\"),\n",
    "    api_key=os.getenv(\"OPENAI_API_KEY\"),\n",
    "    model=os.getenv(\"MODEL_NAME\"),\n",
    "    streaming=True,\n",
    "    http_async_client=httpx.AsyncClient(verify=False)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "7737dce6-17b4-417d-93eb-2e509e74bbc2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "chain = (\n",
    "    prompt \n",
    "    | llm_reasoning \n",
    "    | parser\n",
    ")\n",
    "\n",
    "\n",
    "response = await chain.ainvoke(batch_inputs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "001ea29d-ca16-4c35-9b8b-8507ecd6e758",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"criterias\": [\n",
      "        {\n",
      "            \"decision\": \"NO\",\n",
      "            \"rationale\": \"The title does not mention 'colorectal cancer'. It refers to hepatocellular carcinoma.\"\n",
      "        },\n",
      "        {\n",
      "            \"decision\": \"YES\",\n",
      "            \"rationale\": \"The term 'crizotinib' is mentioned in the provided paper content.\"\n",
      "        },\n",
      "        {\n",
      "            \"decision\": \"NO\",\n",
      "            \"rationale\": \"There is no comparison made between crizotinib and another treatment in the provided content.\"\n",
      "        },\n",
      "        {\n",
      "            \"decision\": \"YES\",\n",
      "            \"rationale\": \"Methods and results sections are described in the paper, although details might be limited.\"\n",
      "        },\n",
      "        {\n",
      "            \"decision\": \"NO\",\n",
      "            \"rationale\": \"The paper does not report specific outcomes such as overall survival, progression-free survival, response rate, or adverse events.\"\n",
      "        },\n",
      "        {\n",
      "            \"decision\": \"UNCERTAIN\",\n",
      "            \"rationale\": \"It is unclear whether the study design is an RCT, prospective cohort study, or retrospective case-control study from the provided information.\"\n",
      "        }\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(response.model_dump_json(indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3d5eb6c-4679-4ac6-8669-7a8aae5a0c3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "response = await chain.ainvoke({\"text\": text, \"fmt\": parser.get_format_instructions()})\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "9ca40aaa-8cfb-4952-9def-dcefef9a38d6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Unsupported template format: [{'P': ('Patients with Colorectal cancer undergoing treatment with Crizotinib',), 'I': ('Crizotinib',), 'C': ('',), 'O': ('',), 'paper_content': (((['Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC.\\nWe present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance.\\nNGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts.\\nOur findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.'],),),), 'criteria_text': [\"Does the title mention 'colorectal cancer'?\", \"Does the title include 'crizotinib' or similar terms like 'Crizolta', 'PF-02341066'?\", 'Is there any indication in the title that the study compares crizotinib to another treatment?', 'Are the methods section and results clearly described in the paper?', 'Do the results specifically report outcomes such as overall survival, progression-free survival, response rate, or adverse events?', 'Was the study randomized controlled trial (RCT), prospective cohort study, or retrospective case-control study? — Preferably RCTs for higher reliability.'], 'num_criteria': 6}]",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[83], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mChatPromptTemplate\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_messages\u001b[49m\u001b[43m(\u001b[49m\u001b[43mLITERATURE_SCREENING_FC\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbatch_inputs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\langchain_core\\prompts\\chat.py:1167\u001b[0m, in \u001b[0;36mChatPromptTemplate.from_messages\u001b[1;34m(cls, messages, template_format)\u001b[0m\n\u001b[0;32m   1118\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[0;32m   1119\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mfrom_messages\u001b[39m(\n\u001b[0;32m   1120\u001b[0m     \u001b[38;5;28mcls\u001b[39m,\n\u001b[0;32m   1121\u001b[0m     messages: Sequence[MessageLikeRepresentation],\n\u001b[0;32m   1122\u001b[0m     template_format: PromptTemplateFormat \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mf-string\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   1123\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ChatPromptTemplate:\n\u001b[0;32m   1124\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Create a chat prompt template from a variety of message formats.\u001b[39;00m\n\u001b[0;32m   1125\u001b[0m \n\u001b[0;32m   1126\u001b[0m \u001b[38;5;124;03m    Examples:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1165\u001b[0m \n\u001b[0;32m   1166\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m-> 1167\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mmessages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemplate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtemplate_format\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\langchain_core\\prompts\\chat.py:971\u001b[0m, in \u001b[0;36mChatPromptTemplate.__init__\u001b[1;34m(self, messages, template_format, **kwargs)\u001b[0m\n\u001b[0;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m__init__\u001b[39m(\n\u001b[0;32m    903\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m    904\u001b[0m     messages: Sequence[MessageLikeRepresentation],\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    907\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[0;32m    908\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    909\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Create a chat prompt template from a variety of message formats.\u001b[39;00m\n\u001b[0;32m    910\u001b[0m \n\u001b[0;32m    911\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    969\u001b[0m \u001b[38;5;124;03m        ```\u001b[39;00m\n\u001b[0;32m    970\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 971\u001b[0m     messages_ \u001b[38;5;241m=\u001b[39m [\n\u001b[0;32m    972\u001b[0m         _convert_to_message_template(message, template_format)\n\u001b[0;32m    973\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m message \u001b[38;5;129;01min\u001b[39;00m messages\n\u001b[0;32m    974\u001b[0m     ]\n\u001b[0;32m    976\u001b[0m     \u001b[38;5;66;03m# Automatically infer input variables from messages\u001b[39;00m\n\u001b[0;32m    977\u001b[0m     input_vars: \u001b[38;5;28mset\u001b[39m[\u001b[38;5;28mstr\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m()\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\langchain_core\\prompts\\chat.py:972\u001b[0m, in \u001b[0;36m<listcomp>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21m__init__\u001b[39m(\n\u001b[0;32m    903\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m    904\u001b[0m     messages: Sequence[MessageLikeRepresentation],\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    907\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[0;32m    908\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    909\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Create a chat prompt template from a variety of message formats.\u001b[39;00m\n\u001b[0;32m    910\u001b[0m \n\u001b[0;32m    911\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    969\u001b[0m \u001b[38;5;124;03m        ```\u001b[39;00m\n\u001b[0;32m    970\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m    971\u001b[0m     messages_ \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m--> 972\u001b[0m         \u001b[43m_convert_to_message_template\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmessage\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemplate_format\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    973\u001b[0m         \u001b[38;5;28;01mfor\u001b[39;00m message \u001b[38;5;129;01min\u001b[39;00m messages\n\u001b[0;32m    974\u001b[0m     ]\n\u001b[0;32m    976\u001b[0m     \u001b[38;5;66;03m# Automatically infer input variables from messages\u001b[39;00m\n\u001b[0;32m    977\u001b[0m     input_vars: \u001b[38;5;28mset\u001b[39m[\u001b[38;5;28mstr\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m()\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\langchain_core\\prompts\\chat.py:1439\u001b[0m, in \u001b[0;36m_convert_to_message_template\u001b[1;34m(message, template_format)\u001b[0m\n\u001b[0;32m   1437\u001b[0m     message_ \u001b[38;5;241m=\u001b[39m message\n\u001b[0;32m   1438\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(message, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m-> 1439\u001b[0m     message_ \u001b[38;5;241m=\u001b[39m \u001b[43m_create_template_from_message_type\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1440\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mhuman\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmessage\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemplate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtemplate_format\u001b[49m\n\u001b[0;32m   1441\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1442\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(message, (\u001b[38;5;28mtuple\u001b[39m, \u001b[38;5;28mdict\u001b[39m)):\n\u001b[0;32m   1443\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(message, \u001b[38;5;28mdict\u001b[39m):\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\langchain_core\\prompts\\chat.py:1349\u001b[0m, in \u001b[0;36m_create_template_from_message_type\u001b[1;34m(message_type, template, template_format)\u001b[0m\n\u001b[0;32m   1335\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Create a message prompt template from a message type and template string.\u001b[39;00m\n\u001b[0;32m   1336\u001b[0m \n\u001b[0;32m   1337\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1346\u001b[0m \u001b[38;5;124;03m    ValueError: If unexpected message type.\u001b[39;00m\n\u001b[0;32m   1347\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m   1348\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m message_type \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhuman\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muser\u001b[39m\u001b[38;5;124m\"\u001b[39m}:\n\u001b[1;32m-> 1349\u001b[0m     message: BaseMessagePromptTemplate \u001b[38;5;241m=\u001b[39m \u001b[43mHumanMessagePromptTemplate\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_template\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1350\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtemplate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemplate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtemplate_format\u001b[49m\n\u001b[0;32m   1351\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1352\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m message_type \u001b[38;5;129;01min\u001b[39;00m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mai\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124massistant\u001b[39m\u001b[38;5;124m\"\u001b[39m}:\n\u001b[0;32m   1353\u001b[0m     message \u001b[38;5;241m=\u001b[39m AIMessagePromptTemplate\u001b[38;5;241m.\u001b[39mfrom_template(\n\u001b[0;32m   1354\u001b[0m         cast(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstr\u001b[39m\u001b[38;5;124m\"\u001b[39m, template), template_format\u001b[38;5;241m=\u001b[39mtemplate_format\n\u001b[0;32m   1355\u001b[0m     )\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\langchain_core\\prompts\\chat.py:437\u001b[0m, in \u001b[0;36m_StringImageMessagePromptTemplate.from_template\u001b[1;34m(cls, template, template_format, partial_variables, **kwargs)\u001b[0m\n\u001b[0;32m    419\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Create a class from a string template.\u001b[39;00m\n\u001b[0;32m    420\u001b[0m \n\u001b[0;32m    421\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    434\u001b[0m \u001b[38;5;124;03m    ValueError: If the template is not a string or list of strings.\u001b[39;00m\n\u001b[0;32m    435\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    436\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(template, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m--> 437\u001b[0m     prompt: StringPromptTemplate \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mlist\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[43mPromptTemplate\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_template\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    438\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtemplate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    439\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtemplate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtemplate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    440\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpartial_variables\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpartial_variables\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    441\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    442\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mcls\u001b[39m(prompt\u001b[38;5;241m=\u001b[39mprompt, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m    443\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(template, \u001b[38;5;28mlist\u001b[39m):\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\langchain_core\\prompts\\prompt.py:298\u001b[0m, in \u001b[0;36mPromptTemplate.from_template\u001b[1;34m(cls, template, template_format, partial_variables, **kwargs)\u001b[0m\n\u001b[0;32m    256\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[0;32m    257\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mfrom_template\u001b[39m(\n\u001b[0;32m    258\u001b[0m     \u001b[38;5;28mcls\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    263\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[0;32m    264\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m PromptTemplate:\n\u001b[0;32m    265\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Load a prompt template from a template.\u001b[39;00m\n\u001b[0;32m    266\u001b[0m \n\u001b[0;32m    267\u001b[0m \u001b[38;5;124;03m    !!! warning \"Security\"\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    296\u001b[0m \u001b[38;5;124;03m        The prompt template loaded from the template.\u001b[39;00m\n\u001b[0;32m    297\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 298\u001b[0m     input_variables \u001b[38;5;241m=\u001b[39m \u001b[43mget_template_variables\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtemplate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemplate_format\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    299\u001b[0m     partial_variables_ \u001b[38;5;241m=\u001b[39m partial_variables \u001b[38;5;129;01mor\u001b[39;00m {}\n\u001b[0;32m    301\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m partial_variables_:\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\langchain_core\\prompts\\string.py:280\u001b[0m, in \u001b[0;36mget_template_variables\u001b[1;34m(template, template_format)\u001b[0m\n\u001b[0;32m    278\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    279\u001b[0m     msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnsupported template format: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtemplate_format\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m--> 280\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[0;32m    282\u001b[0m \u001b[38;5;66;03m# For f-strings, block attribute access and indexing syntax\u001b[39;00m\n\u001b[0;32m    283\u001b[0m \u001b[38;5;66;03m# This prevents template injection attacks via accessing dangerous attributes\u001b[39;00m\n\u001b[0;32m    284\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m template_format \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mf-string\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "\u001b[1;31mValueError\u001b[0m: Unsupported template format: [{'P': ('Patients with Colorectal cancer undergoing treatment with Crizotinib',), 'I': ('Crizotinib',), 'C': ('',), 'O': ('',), 'paper_content': (((['Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC.\\nWe present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance.\\nNGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts.\\nOur findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.'],),),), 'criteria_text': [\"Does the title mention 'colorectal cancer'?\", \"Does the title include 'crizotinib' or similar terms like 'Crizolta', 'PF-02341066'?\", 'Is there any indication in the title that the study compares crizotinib to another treatment?', 'Are the methods section and results clearly described in the paper?', 'Do the results specifically report outcomes such as overall survival, progression-free survival, response rate, or adverse events?', 'Was the study randomized controlled trial (RCT), prospective cohort study, or retrospective case-control study? — Preferably RCTs for higher reliability.'], 'num_criteria': 6}]"
     ]
    }
   ],
   "source": [
    "ChatPromptTemplate.from_messages(LITERATURE_SCREENING_FC, batch_inputs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "41e951a0-0e81-4bb1-8bc8-e09d65728f3e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from langchain_openai import ChatOpenAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "409a4971-4f7c-4a5c-94dd-0b22e9da5290",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "```json\n",
      "{\n",
      "    \"events\": [\n",
      "        {\n",
      "            \"date\": \"2025-11-04\",\n",
      "            \"time\": \"14:00\",\n",
      "            \"reason\": \"watching the movie 'The Dark Knight'\",\n",
      "            \"title\": \"Cinema Visit\",\n",
      "            \"description\": \"Going to the cinema to watch the movie 'The Dark Knight'.\"\n",
      "        },\n",
      "        {\n",
      "            \"date\": \"2025-11-04\",\n",
      "            \"time\": \"16:00\",\n",
      "            \"reason\": \"dinner after the movie\",\n",
      "            \"title\": \"Dinner After Movie\",\n",
      "            \"description\": \"Having dinner in a nearby restaurant after watching the movie.\"\n",
      "        }\n",
      "    ]\n",
      "}\n",
      "```\n"
     ]
    }
   ],
   "source": [
    "prompt = ChatPromptTemplate.from_messages([\n",
    "    (\n",
    "        \"system\", \n",
    "        dedent(\n",
    "            \"\"\"\n",
    "            You are a helpful assistant. \n",
    "            Your goal is to extract events related information from the text provided by the user.\n",
    "            Return the information in the following JSON-format.\n",
    "            ```json\n",
    "            {{\n",
    "                \"events\": [\n",
    "                    {{\n",
    "                        \"date\": \"date of the event\",\n",
    "                        \"time\": \"time of the event\",\n",
    "                        \"reason\": \"why the activity was chosen\"\n",
    "                        \"title\": \"title of the event\",\n",
    "                        \"description\": \"description of the event\"\n",
    "                    }}\n",
    "                ]\n",
    "            }}\n",
    "            ```\n",
    "            if there is no events, return an empty list.\n",
    "            /no_think\n",
    "            \"\"\"\n",
    "        )\n",
    "    ),\n",
    "    (\n",
    "        \"human\", \n",
    "        dedent(\n",
    "            \"\"\"\n",
    "            Text: {text}\"\n",
    "            \"\"\"\n",
    "        )\n",
    "    ),\n",
    "])\n",
    "\n",
    "chain = (\n",
    "    prompt \n",
    "    | llm_reasoning \n",
    "    | StrOutputParser()\n",
    ")\n",
    "\n",
    "\n",
    "text = \"\"\"\n",
    "    We are planning to go to the cinema on 2025-11-04 at 14-00 to watch the movie 'The Dark Knight'. \n",
    "    Probably, we also will have dinner after the movie in the nearby restaurant two hours later.\n",
    "\"\"\"\n",
    "response = await chain.ainvoke({\"text\": text})\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e785c6e-737b-4f76-9cdf-d8dce4edf47a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4442059-f9e6-4373-a9f5-a1c9a9932a7d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "108da99a-9c81-4cde-bc04-89ac3431840a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d71c255e-1f6c-429c-8e14-3c03f724a1bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44cefd9e-0064-4e35-bc8a-b3bb9838fc4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c69ff7c4-4e91-4afc-bdae-ec1fbf094012",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/41268440/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/41213063/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40526090/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40211189/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40369167/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n"
     ]
    }
   ],
   "source": [
    "pmid_list = df_papers[0].PMID.values.tolist()\n",
    "#['41213063',#'26451310',]\n",
    "\n",
    "res = pmid2biocxml(pmid_list, api_key=pubmed_api_key)\n",
    "#print(res)\n",
    "res = [parse_bioc_xml(r) for r in res]\n",
    "# transform the parsed xml into paper content\n",
    "papers = []\n",
    "for parsed in res:\n",
    "    paper_content = []\n",
    "    for parsed_ in parsed[\"passage\"]:\n",
    "        paper_content.append(parsed_['content'])\n",
    "    paper_content = \"\\n\".join(paper_content)\n",
    "    papers.append(paper_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ac1dabb-884d-4bd7-9ce3-edd08f6c9e35",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1672, 1652, 27171, 2627, 46819]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[len(i) for i in papers]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6a9ab051-ac79-4050-93d9-7008b5074ed6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.',\n",
       " 'Background: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic',\n",
       " 'challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,',\n",
       " 'analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts. Conclusion: Our',\n",
       " 'a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts. Conclusion: Our findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to',\n",
       " 'Our findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splited = split_text_into_chunks(papers[1], chunk_size=600,\n",
    "                                             chunk_overlap=300)\n",
    "splited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ac23d300-fc62-407f-8265-409a50f8a216",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[{'name': 'Crizotinib treatment effectiveness',\n",
       "   'value': 'initial tumor regression followed by disease progression',\n",
       "   'source_id': [2],\n",
       "   'cited_blocks': ['challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,']}],\n",
       " [{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'More sensitive to low-risk group',\n",
       "   'source_id': [5],\n",
       "   'cited_blocks': ['performed in risk groups. Expression of signature genes in CRC cells was examined using quantitative reverse transcription polymerase chain reaction (qRT-PCR). There were 10 CRC-related RRAGs found. Prognostic models were built with RRAGs. Based on immune infiltration analysis, low-risk populations showed significantly greater levels of immune infiltration (P P ARHGEF17 mutation rate was 6%. Medications such as CGP-082996, Dasatinib, Erlotinib, and Salubrinal were more sensitive to high-risk group, whereas drugs such as FTI-277, DMOG, and Crizotinib were more sensitive to low-risk group.']}],\n",
       " [{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'NP',\n",
       "   'source_id': [],\n",
       "   'cited_blocks': []}],\n",
       " [{'name': 'main_results_crizotinib',\n",
       "   'value': 'no objective responses observed in RASMT advanced CRC patients',\n",
       "   'source_id': [7],\n",
       "   'cited_blocks': ['with low baselineÂ\\xa0KRASMT allele frequency. CONCLUSIONS: Combination binimetinib/crizotinib showed a poor tolerability with no objective responses observed in RASMT advanced CRC patients. EudraCT-Number: 2014-000463 -â\\x80\\x8940 (20/06/2014: A Sequential Phase I study of MEK1/2 inhibitors PD- 0325901 or Binimetinib combined with cMET inhibitor Crizotinib in RAS Mutant and RAS Wild Type with aberrant c-MET).']}],\n",
       " [{'name': 'Crizotinib_effectiveness',\n",
       "   'value': 'selectively inhibited the growth of ARID1A-deficient CRC cells',\n",
       "   'source_id': [80, 142],\n",
       "   'cited_blocks': ['validated the synthetic lethal effect of the c-MET inhibitor PHA-665752 (PHA) in two ARID1A knockout (ARID1A-KO) clones, where PHA selectively inhibited the growth of ARID1A-KO HCT116 cells (Fig.Â\\xa01D). Moreover, two FDA-approved c-MET inhibitors, crizotinib, and cabozantinib, also showed synthetic lethality in ARID1A-KO HCT116 clones (Fig.Â\\xa01E, F). These results suggest that c-MET is a potential synthetic lethal target for ARID1A in CRC cells.',\n",
       "    'synthetic lethality in ARID1A-deficient CRC. Notably, FDA-approved c-MET inhibitors, such as crizotinib and cabozantinib, have demonstrated enhanced cytotoxicity in ARID1A-deficient cancers, facilitating the translation of these approved drugs for repurposing in ARID1A-targeted precision cancer medicine.']}]]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "api = StudyCharacteristicsExtraction()\n",
    "extracted = api.run(\n",
    "    papers=papers,\n",
    "    fields=['Results, string, main results of the study regarding Crizotinib, usually about treatment effectiveness'],#api.DEFAULT_FIELDS,\n",
    "    llm='Qwen/Qwen3-32B',#\"openai-gpt-4o\",\n",
    "    chunk_size=600,\n",
    "    chunk_overlap=300,\n",
    "    thinking=False,\n",
    ")\n",
    "extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "dc85b6da-8fe9-4484-8430-e55e239b2525",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'More sensitive to low-risk group',\n",
       "   'source_id': [4],\n",
       "   'cited_blocks': ['performed in risk groups. Expression of signature genes in CRC cells was examined using quantitative reverse transcription polymerase chain reaction (qRT-PCR). There were 10 CRC-related RRAGs found. Prognostic models were built with RRAGs. Based on immune infiltration analysis, low-risk populations showed significantly greater levels of immune infiltration (P P ARHGEF17 mutation rate was 6%. Medications such as CGP-082996, Dasatinib, Erlotinib, and Salubrinal were more sensitive to high-risk group, whereas drugs such as FTI-277, DMOG, and Crizotinib were more sensitive to low-risk group.']}],\n",
       " [{'name': 'Crizotinib treatment effectiveness',\n",
       "   'value': 'initial tumor regression followed by disease progression',\n",
       "   'source_id': [1],\n",
       "   'cited_blocks': ['challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,']}]]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "80601546-e32e-4b65-b837-b5dec0abdfc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm_reasoning = ChatOpenAI(\n",
    "    base_url=BASE_URL,\n",
    "    api_key=HUGGINGFACE_HUB_TOKEN,\n",
    "    model=MODEL_NAME_HF,\n",
    "    streaming=True,\n",
    "    temperature=0,\n",
    "    #http_client=httpx_client\n",
    "    http_async_client=httpx.AsyncClient(verify=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "87efd589-f082-4326-bc29-42ab4cd01c9a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<think>\n",
      "\n",
      "</think>\n",
      "\n",
      "```json\n",
      "[\n",
      "    {\n",
      "        \"name\": \"Crizotinib effectiveness\",\n",
      "        \"value\": \"More sensitive to low-risk group\",\n",
      "        \"source_id\": [3]\n",
      "    }\n",
      "]\n",
      "```\n"
     ]
    }
   ],
   "source": [
    "prompt = ChatPromptTemplate.from_messages([\n",
    "    (\"user\", '/no_think {text}'),\n",
    "    # (\"ai\", \"Good.\")\n",
    "])\n",
    "\n",
    "chain = prompt | llm_reasoning\n",
    "response = await chain.ainvoke({\"text\": ff})\n",
    "\n",
    "#fin_condition = response.content.strip('<think>\\n\\n</think>\\n\\n')\n",
    "print(response.content)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:llms]",
   "language": "python",
   "name": "conda-env-llms-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
