{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "26194509-66fb-4e01-b3c6-47c7952a13c3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv, find_dotenv\n",
    "\n",
    "load_dotenv(find_dotenv(usecwd=True))\n",
    "\n",
    "BASE_URL = \"https://gigachat.devices.sberbank.ru/api/v1\"\n",
    "#base_gg = \"https://router.huggingface.co/v1\"\n",
    "#\"http://a6k2.dgx:34000/v1\"\n",
    "\n",
    "BASE_URL = \"https://router.huggingface.co/v1\"\n",
    "\n",
    "#\"qwen3-32b\"\n",
    "# gigachat/GigaChat\n",
    "MODEL_NAME = os.getenv(\"MODEL_NAME\", 'GigaChat')\n",
    "\n",
    "\n",
    "\n",
    "MODEL_NAME_HF ='meta-llama/Llama-3.3-70B-Instruct'\n",
    "MODEL_NAME_HF ='Qwen/Qwen3-32B'\n",
    "\n",
    "MODEL_TEMPERATURE_HF =1\n",
    "\n",
    "os.environ[\"OPENAI_API_KEY\"] = os.getenv(\"GIGACHAT_AUTH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a43a81f3-e83a-4a63-89a6-bd0cc608eff0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MDE5YmM2MjctOWU5ZS03OWE0LWE0OGYtY2FhMTU4MzYxZDgzOjg2MjAwMDgzLThkNzMtNGU0ZS1hNzIxLTM3YWY5ZWY4ZmZlNg=='"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getenv(\"GIGACHAT_AUTH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e14f07d4-cfbd-4129-8f11-5bf5c74d31cf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from langchain_community.document_loaders import PyMuPDFLoader\n",
    "from pydantic import BaseModel, Field\n",
    "from langchain_core.output_parsers import PydanticOutputParser, StrOutputParser, BaseOutputParser\n",
    "from textwrap import dedent\n",
    "\n",
    "from langchain_openai import ChatOpenAI\n",
    "from langchain_core.messages import HumanMessage\n",
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "import httpx\n",
    "import json\n",
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "from langchain_pymupdf4llm import PyMuPDF4LLMLoader, PyMuPDF4LLMParser\n",
    "import pymupdf4llm\n",
    "\n",
    "from trialmind.pubmed import pmid2papers, PubmedAPIWrapper, pmid2biocxml, parse_bioc_xml\n",
    "from trialmind.api import StudyCharacteristicsExtraction, ScreeningCriteriaGeneration\n",
    "from trialmind.retrievers import split_text_into_chunks\n",
    "\n",
    "\n",
    "import time\n",
    "import extract\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5301994a-a15c-47ce-b6b8-ce0ceae9e67b",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### clinical trials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7eb1ed4e-88ad-454d-a841-868d8d03925b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Colorectal cancer',\n",
       " ['Crizotinib',\n",
       "  'Lorlatinib',\n",
       "  'Necitumumab',\n",
       "  'Nimotuzumab',\n",
       "  'Panitumumab',\n",
       "  'Cetuximab',\n",
       "  'Brigatinib',\n",
       "  'Ramucirumab',\n",
       "  'Pomalidomide',\n",
       "  'Toremifene',\n",
       "  'Denosumab',\n",
       "  'Duvelisib',\n",
       "  'Inavolitib',\n",
       "  'Bevacizumab',\n",
       "  'Anastrozole',\n",
       "  'Exemestane'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"docs/CC_215_L00_DL_fusion_fixed_oncobox_ru.pdf\"\n",
    "fin_condition, treatements_eng = extract.info_from_doc(file_path)\n",
    "fin_condition, treatements_eng"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3c68437-54be-4611-9aa0-d7f6a379a924",
   "metadata": {},
   "source": [
    "1. загрузить док\n",
    "2. взять 1-ую страницу; рег.выражениями взять диагноз\n",
    "-- с ЛЛМ перевести на англ.\n",
    "3. найти страницу с таблицей; или взять таблицы с PyMuPDFLoader\n",
    "4. извлечь таблицу"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "e04c59f8-3cbd-4491-9154-56edbc434e12",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Crizotinib OR Lorlatinib OR Necitumumab OR Nimotuzumab OR Panitumumab OR Cetuximab OR Brigatinib OR Ramucirumab OR Pomalidomide OR Toremifene OR Denosumab OR Duvelisib OR Inavolitib OR Bevacizumab OR Anastrozole OR Exemestane'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "' OR '.join(treatements_eng)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6b530143-823c-495a-8e2a-7039031a23b2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0b55cf5c-e883-4a42-ad8f-08633873be8f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_studies = extract.get_clinicaltrials(fin_condition, \n",
    "                                      #' OR '.join(treatements_eng), \n",
    "                                       treatements_eng[0],  \n",
    "                                      max_studies=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30556d38-abf1-4810-a15e-46d3eef966dc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>results</th>\n",
       "      <th>id</th>\n",
       "      <th>title</th>\n",
       "      <th>status</th>\n",
       "      <th>conditions</th>\n",
       "      <th>study_type</th>\n",
       "      <th>phases</th>\n",
       "      <th>interventions</th>\n",
       "      <th>outcomes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT02465060</td>\n",
       "      <td>Targeted Therapy Directed by Genetic Testing i...</td>\n",
       "      <td>ACTIVE_NOT_RECRUITING</td>\n",
       "      <td>[Advanced Lymphoma, Advanced Malignant Solid N...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[adavosertib, afatinib, afatinib dimaleate, bi...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT03947385</td>\n",
       "      <td>Study of IDE196 in Patients With Solid Tumors ...</td>\n",
       "      <td>RECRUITING</td>\n",
       "      <td>[Metastatic Uveal Melanoma, Cutaneous Melanoma...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE1, PHASE2]</td>\n",
       "      <td>[ide196, binimetinib, crizotinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT02568267</td>\n",
       "      <td>Basket Study of Entrectinib (RXDX-101) for the...</td>\n",
       "      <td>ACTIVE_NOT_RECRUITING</td>\n",
       "      <td>[Breast Cancer, Cholangiocarcinoma, Colorectal...</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[entrectinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT06214793</td>\n",
       "      <td>Taletrectinib in Previously Treated Metastatic...</td>\n",
       "      <td>SUSPENDED</td>\n",
       "      <td>[Breast Cancer, Metastatic Breast Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[taletrectinib]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT03792568</td>\n",
       "      <td>ALK Inhibitor in Metastatic Colorectal Cancer ...</td>\n",
       "      <td>UNKNOWN</td>\n",
       "      <td>[Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[NA]</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>False</td>\n",
       "      <td>NCT05725200</td>\n",
       "      <td>Study to Investigate Outcome of Individualized...</td>\n",
       "      <td>RECRUITING</td>\n",
       "      <td>[Metastatic Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE2]</td>\n",
       "      <td>[alectinib, cetuximab, crizotinib, dasatinib, ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>NCT02510001</td>\n",
       "      <td>MEK and MET Inhibition in Colorectal Cancer</td>\n",
       "      <td>COMPLETED</td>\n",
       "      <td>[Solid Tumor, Colorectal Cancer]</td>\n",
       "      <td>INTERVENTIONAL</td>\n",
       "      <td>[PHASE1]</td>\n",
       "      <td>[pf-02341066, pd-0325901, binimetinib]</td>\n",
       "      <td>[{'type': 'PRIMARY', 'title': 'Maximal Tolerat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   results           id                                              title  \\\n",
       "0    False  NCT02465060  Targeted Therapy Directed by Genetic Testing i...   \n",
       "1    False  NCT03947385  Study of IDE196 in Patients With Solid Tumors ...   \n",
       "2    False  NCT02568267  Basket Study of Entrectinib (RXDX-101) for the...   \n",
       "3    False  NCT06214793  Taletrectinib in Previously Treated Metastatic...   \n",
       "4    False  NCT03792568  ALK Inhibitor in Metastatic Colorectal Cancer ...   \n",
       "5    False  NCT05725200  Study to Investigate Outcome of Individualized...   \n",
       "6     True  NCT02510001        MEK and MET Inhibition in Colorectal Cancer   \n",
       "\n",
       "                  status                                         conditions  \\\n",
       "0  ACTIVE_NOT_RECRUITING  [Advanced Lymphoma, Advanced Malignant Solid N...   \n",
       "1             RECRUITING  [Metastatic Uveal Melanoma, Cutaneous Melanoma...   \n",
       "2  ACTIVE_NOT_RECRUITING  [Breast Cancer, Cholangiocarcinoma, Colorectal...   \n",
       "3              SUSPENDED          [Breast Cancer, Metastatic Breast Cancer]   \n",
       "4                UNKNOWN                                [Colorectal Cancer]   \n",
       "5             RECRUITING                     [Metastatic Colorectal Cancer]   \n",
       "6              COMPLETED                   [Solid Tumor, Colorectal Cancer]   \n",
       "\n",
       "       study_type            phases  \\\n",
       "0  INTERVENTIONAL          [PHASE2]   \n",
       "1  INTERVENTIONAL  [PHASE1, PHASE2]   \n",
       "2  INTERVENTIONAL          [PHASE2]   \n",
       "3  INTERVENTIONAL          [PHASE2]   \n",
       "4  INTERVENTIONAL              [NA]   \n",
       "5  INTERVENTIONAL          [PHASE2]   \n",
       "6  INTERVENTIONAL          [PHASE1]   \n",
       "\n",
       "                                       interventions  \\\n",
       "0  [adavosertib, afatinib, afatinib dimaleate, bi...   \n",
       "1                  [ide196, binimetinib, crizotinib]   \n",
       "2                                      [entrectinib]   \n",
       "3                                    [taletrectinib]   \n",
       "4                                                 []   \n",
       "5  [alectinib, cetuximab, crizotinib, dasatinib, ...   \n",
       "6             [pf-02341066, pd-0325901, binimetinib]   \n",
       "\n",
       "                                            outcomes  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2                                                NaN  \n",
       "3                                                NaN  \n",
       "4                                                NaN  \n",
       "5                                                NaN  \n",
       "6  [{'type': 'PRIMARY', 'title': 'Maximal Tolerat...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_studies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37f1ccfb-f682-457e-bf61-1bed5aad563c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### outcome results\n",
    "Взяли все результаты, отдельно сохранили испытания с результатами, и з них отбираем рез-ты только с unit==participants\n",
    "\n",
    "filter by interventional study type?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "c5e73eef-7cac-4115-93c7-34ba0ee46532",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'type': 'PRIMARY',\n",
       "  'title': 'Clinical Response to Binimetinib Combined With PF-02341066',\n",
       "  'description': 'To investigate response to treatment with RPII dose of Binimetinib with Crizotinib (PF-02341066), in patients with a) RASMT CRC or b) RASWT/cMET mut amplified CRC or c) RASWT/c-MET over-expressed CRC, as defined by stable, partially or completely responding disease, per Response Evaluation Criteria In Solid Tumors Criteria (RECIST v1.1) for target lesions and assessed by CT: Complete Response (CR), Disappearance of all target lesions; Partial Response (PR), \\\\>=30% decrease in the sum of the longest diameter of target lesions; Stable Disease (SD), Neither sufficient shrinkage to qualify for PR nor sufficient increase (\\\\>=20%) to qualify for Progressive Disease; Overall Response (OR) = CR + PR + SD',\n",
       "  'populationDescription': 'Evaluable patients for the primary outcome are those patients who complete a response assessment after cycle 1 of treatment, or who progress early on treatment. 30 of the 36 recruited patients had a response assessment after cycle 1 or progressed early on treatment, and hence these 30 patients are evaluable for the primary analysis.',\n",
       "  'reportingStatus': 'POSTED',\n",
       "  'paramType': 'COUNT_OF_PARTICIPANTS',\n",
       "  'unitOfMeasure': 'Participants',\n",
       "  'timeFrame': 'Dose Expansion phase: change from baseline and up to 12 months.',\n",
       "  'groups': [{'id': 'OG000',\n",
       "    'title': 'Dose Expansion Phase',\n",
       "    'description': 'Binimetinib 30mg BD interval dose administration Days 1-21 every 28 days PF-02341066 (Crizotinib) 250mg OD Days 1-28 continuously Dosage determined following the recommended Phase II dose identification in the dose escalation phase.\\n\\nPF-02341066: PF-02341066 (Crizotinib) 250mg OD or 200mg BD or 250mg BD Days 1-28 continuously\\n\\nBinimetinib: Binimetinib 30mg or 45 mg BD either continuously or Days 1-21 every 28 days'}],\n",
       "  'denoms': [{'units': 'Participants',\n",
       "    'counts': [{'groupId': 'OG000', 'value': '30'}]}],\n",
       "  'classes': [{'categories': [{'title': 'Stable Disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '7'}]},\n",
       "     {'title': 'Progressive Disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '22'}]},\n",
       "     {'title': 'Early death from malignant disease',\n",
       "      'measurements': [{'groupId': 'OG000', 'value': '1'}]}]}]}]"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outcome_participants = [obj for item_main in results_studies for obj in item_main['resultsSection']['outcomeMeasuresModule'\n",
    "                                           ]['outcomeMeasures'\n",
    "                                            ] if(obj.get('unitOfMeasure','') == 'Participants')] \n",
    "outcome_participants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9303a48c-9dd0-418b-8c9f-3708b4f9148a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "db6360d3-76fa-4607-8764-0a705af4464c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<think>\n",
      "\n",
      "</think>\n",
      "\n",
      "Here is a concise summary of the efficiency statistics for the clinical response to Binimetinib combined with PF-02341066 (Crizotinib) in the Dose Expansion Phase:\n",
      "\n",
      "- **Total Evaluable Patients**: 30 participants.\n",
      "- **Stable Disease (SD)**: 7 patients (23.3%).\n",
      "- **Progressive Disease (PD)**: 22 patients (73.3%).\n",
      "- **Early Death from Malignant Disease**: 1 patient (3.3%).\n",
      "\n",
      "These outcomes were assessed using RECIST v1.1 criteria after cycle 1 of treatment.\n"
     ]
    }
   ],
   "source": [
    "prompt2 = ChatPromptTemplate.from_messages([\n",
    "    (\"system\", \"You are a helpful medical assistant. Extract information about the efficiency statistics from each outcome measure. Give a concise description with percentages /no_think\"),\n",
    "    (\"human\", \"{text}\"),\n",
    "    # (\"ai\", \"Good.\")\n",
    "])\n",
    "\n",
    "chain2 = prompt2 | llm_reasoning\n",
    "response_res = await chain2.ainvoke({\"text\": outcome_participants[0]})\n",
    "print(response_res.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84f5d879-2920-4ec3-bac4-5ae64fb55099",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e706ee64-c65f-40a9-8606-8e71c97fa5d6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### results in a given format?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f1d0545-8778-43fb-9af8-da1730e79453",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class Info(BaseModel):\n",
    "    diagnose: str = Field(..., description=\"База клинических испытаний\")\n",
    "    treatments: str = Field(..., description=\"Высшие позиции рейтинга таргетных препаратов\")\n",
    "\n",
    "parser = PydanticOutputParser(pydantic_object=Info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d5b0edc-bb8f-4fcd-b443-36745ad7eaf7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "prompt = ChatPromptTemplate.from_messages([\n",
    "    (\n",
    "        \"system\", \n",
    "        dedent(\n",
    "            \"\"\"\n",
    "            You are a helpful assistant. \n",
    "            Your goal is to extract related information from the text provided by the user.\n",
    "            Return the information in the following JSON-format.\n",
    "            {fmt}\n",
    "            \"\"\"\n",
    "        )\n",
    "    ),\n",
    "    (\n",
    "        \"human\", \n",
    "        dedent(\n",
    "            \"\"\"\n",
    "            Text: {text}\"\n",
    "            \"\"\"\n",
    "        )\n",
    "    ),\n",
    "])\n",
    "chain = (\n",
    "    prompt \n",
    "    | llm_reasoning \n",
    "    | parser\n",
    ")\n",
    "\n",
    "\n",
    "response = await chain.ainvoke({\"text\": text, \"fmt\": parser.get_format_instructions()})\n",
    "print(response)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d963b7db-0c01-4b5f-af74-d7d566ae8a54",
   "metadata": {
    "tags": []
   },
   "source": [
    "### papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f1e430e6-c0a6-4a2f-9e92-2b0ca80bdd6e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fin_condition = 'Colorectal cancer'\n",
    "treatements_eng =  ['Crizotinib','Lorlatinib',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ceab5c26-765a-420c-82a6-8bfbacdb36f6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['40526090', '40211189', '41268440', '41213063', '40369167'],\n",
       " 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=(Colorectal cancer)+AND+(Crizotinib)&retmax=5&retmode=json&api_key=6892c4129cef143ff92d11533848d2e0d908',\n",
       " '67')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "search_api = PubmedAPIWrapper()\n",
    "# page_size is the max number of records to return!!!! not pages!\n",
    "tmp_inputs = {\n",
    "        \"page_size\": 5,\n",
    "        \"keyword_map\": {'conditions':[fin_condition], \n",
    "                        'treatments':[treatements_eng[0]]\n",
    "                       },\n",
    "        \"keywords\": {\n",
    "            \"OPERATOR\": 'AND'\n",
    "        }\n",
    "}\n",
    "pubmed_api_key= '6892c4129cef143ff92d11533848d2e0d908'#os.getenv(\"PUBMED_API_KEY\", '')\n",
    "time.sleep(.5)\n",
    "response = search_api.build_search_query_and_get_pmid(tmp_inputs, \n",
    "                                                      api_key=pubmed_api_key)\n",
    "response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a54a4f2c-d723-4a4a-a1b1-77eb50eeaea8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Journal</th>\n",
       "      <th>Year</th>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th>Title</th>\n",
       "      <th>Publication Type</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>40526090</td>\n",
       "      <td>Oncotarget</td>\n",
       "      <td>2025</td>\n",
       "      <td>Jun</td>\n",
       "      <td>17</td>\n",
       "      <td>Case Report WIN-MTB-2023001 WIN International ...</td>\n",
       "      <td>Journal Article, Case Reports</td>\n",
       "      <td>Alberto Hernando-Calvo, Razelle Kurzrock, Nadi...</td>\n",
       "      <td>Heavily pretreated metastatic colorectal cance...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40211189</td>\n",
       "      <td>BMC cancer</td>\n",
       "      <td>2025</td>\n",
       "      <td>Apr</td>\n",
       "      <td>10</td>\n",
       "      <td>A Phase Ia/b study of MEK1/2 inhibitor binimet...</td>\n",
       "      <td>Journal Article, Clinical Trial, Phase I</td>\n",
       "      <td>Francesca Aroldi, Elena Elez, Thierry André, G...</td>\n",
       "      <td>Targeting RAS mutant (MT) colorectal cancer (C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>41268440</td>\n",
       "      <td>Frontiers in pharmacology</td>\n",
       "      <td>2025</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Case Report: HGF and NF1 mutations as putative...</td>\n",
       "      <td>Case Reports, Journal Article</td>\n",
       "      <td>Zhitao Chen, Shan Luo, Yangjun Gu, Qiyong Li</td>\n",
       "      <td>Hepatocellular carcinoma (HCC) is a highly agg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41213063</td>\n",
       "      <td>Journal of environmental pathology, toxicology...</td>\n",
       "      <td>2025</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Radiotherapy-Related Autophagy Genes Predict P...</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Mao Mao, Yichao Zhuang, Yan Chen</td>\n",
       "      <td>Numerous genes have been associated with color...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>40369167</td>\n",
       "      <td>Cell death and differentiation</td>\n",
       "      <td>2025</td>\n",
       "      <td>Nov</td>\n",
       "      <td></td>\n",
       "      <td>ARID1A loss enhances sensitivity to c-MET inhi...</td>\n",
       "      <td>Journal Article</td>\n",
       "      <td>Xu Zhang, Zihuan Wang, Yilin He, Kejin Wang, C...</td>\n",
       "      <td>ARID1A, a subunit of the SWI/SNF chromatin-rem...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                            Journal  Year Month  \\\n",
       "0  40526090                                         Oncotarget  2025   Jun   \n",
       "1  40211189                                         BMC cancer  2025   Apr   \n",
       "2  41268440                          Frontiers in pharmacology  2025         \n",
       "3  41213063  Journal of environmental pathology, toxicology...  2025         \n",
       "4  40369167                     Cell death and differentiation  2025   Nov   \n",
       "\n",
       "  Day                                              Title  \\\n",
       "0  17  Case Report WIN-MTB-2023001 WIN International ...   \n",
       "1  10  A Phase Ia/b study of MEK1/2 inhibitor binimet...   \n",
       "2      Case Report: HGF and NF1 mutations as putative...   \n",
       "3      Radiotherapy-Related Autophagy Genes Predict P...   \n",
       "4      ARID1A loss enhances sensitivity to c-MET inhi...   \n",
       "\n",
       "                           Publication Type  \\\n",
       "0             Journal Article, Case Reports   \n",
       "1  Journal Article, Clinical Trial, Phase I   \n",
       "2             Case Reports, Journal Article   \n",
       "3                           Journal Article   \n",
       "4                           Journal Article   \n",
       "\n",
       "                                             Authors  \\\n",
       "0  Alberto Hernando-Calvo, Razelle Kurzrock, Nadi...   \n",
       "1  Francesca Aroldi, Elena Elez, Thierry André, G...   \n",
       "2       Zhitao Chen, Shan Luo, Yangjun Gu, Qiyong Li   \n",
       "3                   Mao Mao, Yichao Zhuang, Yan Chen   \n",
       "4  Xu Zhang, Zihuan Wang, Yilin He, Kejin Wang, C...   \n",
       "\n",
       "                                            Abstract  \n",
       "0  Heavily pretreated metastatic colorectal cance...  \n",
       "1  Targeting RAS mutant (MT) colorectal cancer (C...  \n",
       "2  Hepatocellular carcinoma (HCC) is a highly agg...  \n",
       "3  Numerous genes have been associated with color...  \n",
       "4  ARID1A, a subunit of the SWI/SNF chromatin-rem...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_papers = pmid2papers(pmid_list=response[0], api_key=pubmed_api_key)\n",
    "df_papers[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dd4f5664-3983-4bbb-af9f-7f460b87e5e3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Colorectal cancer', ['Crizotinib', 'Lorlatinib'])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fin_condition, treatements_eng"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f15bedd0-7b97-4696-ac71-2b9ddc4ef0d6",
   "metadata": {},
   "source": [
    "### extract results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "365da54d-32aa-445c-b27b-0b6b2ae80cf6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fin_condition = 'Colorectal cancer'\n",
    "treatements_eng =  ['Crizotinib','Lorlatinib',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "033c3e2a-28c7-4d05-b135-1078f19e8e3d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MDE5YmM2MjctOWU5ZS03OWE0LWE0OGYtY2FhMTU4MzYxZDgzOjg2MjAwMDgzLThkNzMtNGU0ZS1hNzIxLTM3YWY5ZWY4ZmZlNg=='"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getenv(\"GIGACHAT_AUTH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "66ffb1de-672b-4c10-acfc-bcade11f8c7d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\urllib3\\connectionpool.py:1097: InsecureRequestWarning: Unverified HTTPS request is being made to host 'ngw.devices.sberbank.ru'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'eyJjdHkiOiJqd3QiLCJlbmMiOiJBMjU2Q0JDLUhTNTEyIiwiYWxnIjoiUlNBLU9BRVAtMjU2In0.n3a2FYpWxt9QJhsECw6ggo87K1pT9KrfqJ7AXHi08-i6xR2hFi0noT0lhY3eQ7Nzhv9CskuKrbJhI5y6BiVQEbIJNvWZf6HuhtQp2TWiTlpXe_UC8tRxAhKzJQpkidmHujuKQSRnRugLUgjZZD6YcQaP0wxXeJR4Ae-FGrOrugF5c0oTUi-aXudxHwXDq0h_uyGIHVvz66wnrxVcsE128J8LH0Np7z7TS7P-j0_w3MKZ_7GIOi88lK6gBWNGE2wYivoKI8GB_3c65EC6ItZzrOCqhevWpqBTINAXaRDuXf-EdQG5DFkbqpx4Lse4OwB6XKAsGmwZ1PnLXlrQApgGMw.Eh3pK_ugzcJVwbx7AsWZCw.SnIVvXLeC2FDo2Gf902HrGKrQfYYgsT5-iUKikrIspRAnI3LJcYDaIUpMT_lYK8_W1m8zUtWAzmENmNCVu6hGJyIaAGTuzqDvsgBX0CT7gULHy9mNmEsMqNDsgFtTATdsuYdoB0ZKptb5HoOwlTT-NkcKyWJhsPka_lYe4C5h-4jz7wH6atfjHt5eSuX9ESSWpkREwF2i9tmnRoCVyVUNjr1zune6Qw1FVV4c8LAF6ySJ6sg9CfGktABmRBgCmxbadUGu5CLxuFfhaZZ3CVg73V90VkEb1uqkmnjeA_hijDuX9f1bFt_TWh7ChLileK2WB_W_8JgtM0VfLU6xqs5W329hZdWO17Ctm4diEIH8xe-vgCjnoknsxcKqL4lMTYslscvwLgkQUW8uW5wbCayM0l2YNmW6Z0trwHermNSA5nbRX9v_rPfRAHyJpPsEA5axNfygvoSFTagCuZwRr5lUqjMAI_LTRF2kwG69pm48ULvUZ9RrJ3i4JMqkyTUvW-fzIUebe-h-qurkYBRz7fmT2bDgR4TAB-nqZ2Xa91QyoDo9JbuT7_X0XXpbQKNqkjFNWMNX8cdzONlgkzcNuyiFoqsPQNfIUhZ_WJqnWnXjHaPazER4e1MMSS3FwYmoV2l2jJI6AA7fvOYED2ncn8TG1ACiQmTyZA_niWup0PMrE5Gy-Ieq8t0zOFc5rVlG4BHwVeejZDb4eQIXg8KpBrdcQ9NvsOi8P2Sw9iQLGaSLIE.IxDyTFXQ1vliP2gVhAmyMbee7mEMX103xPh1c_hjerU'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "import uuid\n",
    "\n",
    "url = \"https://ngw.devices.sberbank.ru:9443/api/v2/oauth\"\n",
    "# Создадим идентификатор UUID (36 знаков)\n",
    "rq_uid = str(uuid.uuid4())\n",
    "payload={\n",
    "  'scope': 'GIGACHAT_API_PERS'\n",
    "}\n",
    "headers = {\n",
    "  'Content-Type': 'application/x-www-form-urlencoded',\n",
    "  'Accept': 'application/json',\n",
    "  'RqUID': rq_uid,\n",
    "  'Authorization': f'Basic {os.getenv(\"GIGACHAT_AUTH\")}'\n",
    "}\n",
    "\n",
    "response = requests.request(\"POST\", url, headers=headers, data=payload, verify=False)\n",
    "os.environ[\"OPENAI_API_KEY\"] = response.json()['access_token']\n",
    "\n",
    "os.getenv(\"OPENAI_API_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "720ecef2-268b-4b8e-8241-96b44d26d0bf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from openai import OpenAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4e632a26-387b-4c21-b67c-5408ce327484",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "os.environ[\"BASE_URL\"] = \"https://gigachat.devices.sberbank.ru/api/v1\"\n",
    "os.environ[\"MODEL_NAME\"] ='GigaChat'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "be0895aa-5d22-4c08-b855-21edebeff92f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ChatCompletion(id=None, choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='**Тропинка вдаль бежит,**\\n**Лес зелёный шумит,**\\n**Солнце светит сквозь листву —**\\n**Жизнь прекрасна, вот она!**\\n\\n**Птицы звонко напевают,**\\n**Речка тихо журчит,**\\n**И душа свободна вновь,**\\n**Сердцу мир дарит миг.**', refusal=None, role='assistant', annotations=None, audio=None, function_call=None, tool_calls=None))], created=1770302252, model='GigaChat:2.0.28.2', object='chat.completion', service_tier=None, system_fingerprint=None, usage=CompletionUsage(completion_tokens=76, prompt_tokens=20, total_tokens=96, completion_tokens_details=None, prompt_tokens_details=None, precached_prompt_tokens=2))"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "openai_client = OpenAI(\n",
    "    base_url=os.getenv(\"BASE_URL\"),\n",
    "    api_key=os.getenv(\"OPENAI_API_KEY\"),\n",
    "    http_client=httpx.Client(http2=True, verify=False)\n",
    ")\n",
    "\n",
    "response = openai_client.chat.completions.create(\n",
    "        model=os.getenv(\"MODEL_NAME\"),\n",
    "        messages=[{\"role\": \"user\", \"content\": \"You are a helpful assistant. Create a short poem\"}],  # Ensure messages is a list\n",
    "        temperature=0\n",
    "    )\n",
    "response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "edf59fa4-0dbb-4fd6-8af9-dbaa5190e776",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://gigachat.devices.sberbank.ru/api/v1'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getenv(\"BASE_URL\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f7987e16-3579-4cf1-b9a0-ef66a903aeb9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://gigachat.devices.sberbank.ru/api/v1\n",
      "https://gigachat.devices.sberbank.ru/api/v1/\n",
      "{\n",
      "  \"ELIGIBILITY_ANALYSIS\": [\n",
      "    // Eligibility Criteria Rationales\n",
      "    \"Patients must have colorectal cancer — Studies focusing on other types of cancers would not provide relevant data.\",\n",
      "    \"Studies involving patients receiving crizotinib as an intervention — Exclusion of studies using different treatments ensures consistency in comparison.\",\n",
      "    \"Comparative studies including a control group receiving standard therapy or placebo — Comparisons against standard care or placebo ensure meaningful outcomes.\",\n",
      "    \"Outcomes related to survival rates, response rates, toxicity profiles, or quality-of-life measures — These outcomes directly relate to patient management and treatment effectiveness.\",\n",
      "    \"Published studies from peer-reviewed journals — Ensures reliability and validity of findings.\"\n",
      "  ],\n",
      "  \"TITLE_CRITERIA\": [\n",
      "    // Binary Title-Based Questions\n",
      "    \"Does the title mention 'colorectal cancer'?\",\n",
      "    \"Does the title mention 'crizotinib'?\",\n",
      "    \"Is there any indication in the title that the study compares crizotinib to another treatment?\"\n",
      "  ],\n",
      "  \"CONTENT_CRITERIA\": [\n",
      "    // Binary Content-Based Questions\n",
      "    \"Are the methods section and results clearly described in the paper?\",\n",
      "    \"Does the study specify the type of colorectal cancer (e.g., adenocarcinoma)?\",\n",
      "    \"Were the outcomes measured over a sufficient duration to assess long-term effects?\"\n",
      "  ]\n",
      "}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'title_criteria': [], 'content_criteria': [], 'eligibility_analysis': []}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test the apis\n",
    "\n",
    "api = ScreeningCriteriaGeneration()\n",
    "ecs = api.run(\n",
    "    population = f\"Patients with {fin_condition} undergoing treatment with {treatements_eng[0]}\",\n",
    "    intervention = f\"{treatements_eng[0]}\",\n",
    "    comparator = \"\",\n",
    "    outcome = \"\",\n",
    "    llm = os.getenv(\"MODEL_NAME\"),\n",
    "    num_title_criteria=3,\n",
    "    num_abstract_criteria=3,\n",
    "    thinking=False,\n",
    ")\n",
    "ecs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f2858acb-a1e3-4d89-b8af-f1b7ce0cf1d4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Case Report WIN-MTB-2023001 WIN International Molecular Tumor Board A 62-year-old male with metastatic colorectal cancer with 5 prior lines of treatment.: Heavily pretreated metastatic colorectal cancer (mCRC) poses significant therapeutic challenges. Advances in molecular profiling enables personalized strategies. We present a 62-year-old male with mCRC harboring',\n",
       " 'A Phase Ia/b study of MEK1/2 inhibitor binimetinib with MET inhibitor crizotinib in patients with RAS mutant advanced colorectal cancer (MErCuRIC).: Targeting RAS mutant (MT) colorectal cancer (CRC) remains a difficult challenge, mainly due to the pervasiveness of RAS/MEK-mediated feedback loops. Preclinical studies identified MET/STAT3 as an important mediator of resistance to KRAS-MEK1/2 blockade in RASMT CRC. This dose escalation/expansion study assessed safety and initial efficacy of the MEK1/2 inhibitor binimetinib with MET inhibitor crizotinib in RASMT advanced CRC patients.\\nIn the dose escalation phase, patients with advanced solid tumours received binimetinib with crizotinib, using a rolling- 6 design to determine the maximum tolerable dose (MTD) and safety/tolerability. A subsequent dose expansion in RASMT CRC patients assessed treatment response. Blood samples for pharmacokinetics, MET biomarker and ctDNA analyses, and skin/tumour biopsies for pharmacodynamics, c-MET immunohistochemistry (IHC), MET in situ hybridisation (ISH) and MET DNA-ISH analyses were collected.\\nTwenty patients were recruited in 3 cohorts in the dose escalation. The MTD was binimetinib 30 mg B.D, days 1-21 every 28 days, with crizotinib 250 mg O.D continuously. Dose-limiting toxicities included grade ≥\\u20093 transaminitis, creatinine phosphokinase increases and fatigue. Thirty-six RASMT metastatic CRC patients were enrolled in the dose expansion. Pharmacokinetic and pharmacodynamic parameters showed evidence of target engagement. Across the entire study, the most frequent treatment-related adverse events (TR-AE) were rash (80.4%), fatigue (53.4%) and diarrhoea (51.8%) with grade ≥\\u20093 TR-AE occurring in 44.6%. Best clinical response within the RASMT CRC cohort was stable disease in seven patients (24%). Tumour MET super-expression (IHC H-score >\\u2009180 and MET ISH +\\u20093) was observed in 7 patients (24.1%), with MET-amplification only present in 1 of these patients. This patient discontinued treatment early during cycle 1 due to toxicity. Patients with high baseline\\xa0RASMT allele frequency had a significant shorter median overall survival compared with that seen for patients with low baseline\\xa0KRASMT allele frequency.\\nCombination binimetinib/crizotinib showed a poor tolerability with no objective responses observed in RASMT advanced CRC patients. EudraCT-Number: 2014-000463 -\\u200940 (20/06/2014: A Sequential Phase I study of MEK1/2 inhibitors PD- 0325901 or Binimetinib combined with cMET inhibitor Crizotinib in RAS Mutant and RAS Wild Type with aberrant c-MET).',\n",
       " 'Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC.\\nWe present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance.\\nNGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts.\\nOur findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from trialmind.api import LiteratureScreening\n",
    "api = LiteratureScreening()\n",
    "\n",
    "papers = df_papers[0].iloc[:3][\"Title\"] + \": \" + df_papers[0].iloc[:3][\"Abstract\"].fillna(\"\") # important to fillna\n",
    "papers = papers.tolist()\n",
    "papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d94360-c951-4e9a-b37f-8fc6e2eea203",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e604d682-87c0-4bcb-97c0-0c4109f1bdd3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "title_criteria = [\"Does the title mention 'colorectal cancer'?\",\n",
    "\"Does the title mention 'crizotinib'?\",\n",
    "\"Is there any indication in the title that the study compares crizotinib to another treatment?\"\n",
    "]\n",
    "\n",
    "content_criteria = [\"Are the methods section and results clearly described in the paper?\",\n",
    "\"Does the study specify the type of colorectal cancer (e.g., adenocarcinoma)?\",\n",
    "\"Were the outcomes measured over a sufficient duration to assess long-term effects?\"\n",
    "  ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54842268-da25-4f52-b604-fd1b0473a2e9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6e510df2-8174-4105-bcaa-00767e4560ba",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://gigachat.devices.sberbank.ru/api/v1/\n",
      "https://gigachat.devices.sberbank.ru/api/v1/\n",
      "https://gigachat.devices.sberbank.ru/api/v1/\n"
     ]
    },
    {
     "ename": "RetryError",
     "evalue": "RetryError[<Future at 0x27c8ee3a200 state=finished raised RateLimitError>]",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mRateLimitError\u001b[0m                            Traceback (most recent call last)",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\tenacity\\asyncio\\__init__.py:114\u001b[0m, in \u001b[0;36mAsyncRetrying.__call__\u001b[1;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[0;32m    113\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 114\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m fn(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m    115\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m:  \u001b[38;5;66;03m# noqa: B902\u001b[39;00m\n",
      "File \u001b[1;32m~\\GitHub\\TrialMind-SLR\\trialmind\\llm_utils\\openai_async.py:71\u001b[0m, in \u001b[0;36mapi_function_call_single\u001b[1;34m(client, model, messages, tools, temperature, thinking, **kwargs)\u001b[0m\n\u001b[0;32m     69\u001b[0m     messages[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/no_think \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m+\u001b[39mmessages[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m---> 71\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m client\u001b[38;5;241m.\u001b[39mchat\u001b[38;5;241m.\u001b[39mcompletions\u001b[38;5;241m.\u001b[39mcreate(\n\u001b[0;32m     72\u001b[0m     model\u001b[38;5;241m=\u001b[39mmodel,\n\u001b[0;32m     73\u001b[0m     messages\u001b[38;5;241m=\u001b[39mmessages,\n\u001b[0;32m     74\u001b[0m     tools\u001b[38;5;241m=\u001b[39mtools,\n\u001b[0;32m     75\u001b[0m     temperature\u001b[38;5;241m=\u001b[39mtemperature,\n\u001b[0;32m     76\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs\n\u001b[0;32m     77\u001b[0m )\n\u001b[0;32m     78\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\openai\\resources\\chat\\completions\\completions.py:2678\u001b[0m, in \u001b[0;36mAsyncCompletions.create\u001b[1;34m(self, messages, model, audio, frequency_penalty, function_call, functions, logit_bias, logprobs, max_completion_tokens, max_tokens, metadata, modalities, n, parallel_tool_calls, prediction, presence_penalty, prompt_cache_key, prompt_cache_retention, reasoning_effort, response_format, safety_identifier, seed, service_tier, stop, store, stream, stream_options, temperature, tool_choice, tools, top_logprobs, top_p, user, verbosity, web_search_options, extra_headers, extra_query, extra_body, timeout)\u001b[0m\n\u001b[0;32m   2677\u001b[0m validate_response_format(response_format)\n\u001b[1;32m-> 2678\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_post(\n\u001b[0;32m   2679\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/chat/completions\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m   2680\u001b[0m     body\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mawait\u001b[39;00m async_maybe_transform(\n\u001b[0;32m   2681\u001b[0m         {\n\u001b[0;32m   2682\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessages\u001b[39m\u001b[38;5;124m\"\u001b[39m: messages,\n\u001b[0;32m   2683\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodel\u001b[39m\u001b[38;5;124m\"\u001b[39m: model,\n\u001b[0;32m   2684\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124maudio\u001b[39m\u001b[38;5;124m\"\u001b[39m: audio,\n\u001b[0;32m   2685\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfrequency_penalty\u001b[39m\u001b[38;5;124m\"\u001b[39m: frequency_penalty,\n\u001b[0;32m   2686\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunction_call\u001b[39m\u001b[38;5;124m\"\u001b[39m: function_call,\n\u001b[0;32m   2687\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunctions\u001b[39m\u001b[38;5;124m\"\u001b[39m: functions,\n\u001b[0;32m   2688\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlogit_bias\u001b[39m\u001b[38;5;124m\"\u001b[39m: logit_bias,\n\u001b[0;32m   2689\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlogprobs\u001b[39m\u001b[38;5;124m\"\u001b[39m: logprobs,\n\u001b[0;32m   2690\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax_completion_tokens\u001b[39m\u001b[38;5;124m\"\u001b[39m: max_completion_tokens,\n\u001b[0;32m   2691\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax_tokens\u001b[39m\u001b[38;5;124m\"\u001b[39m: max_tokens,\n\u001b[0;32m   2692\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m\"\u001b[39m: metadata,\n\u001b[0;32m   2693\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodalities\u001b[39m\u001b[38;5;124m\"\u001b[39m: modalities,\n\u001b[0;32m   2694\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mn\u001b[39m\u001b[38;5;124m\"\u001b[39m: n,\n\u001b[0;32m   2695\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparallel_tool_calls\u001b[39m\u001b[38;5;124m\"\u001b[39m: parallel_tool_calls,\n\u001b[0;32m   2696\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprediction\u001b[39m\u001b[38;5;124m\"\u001b[39m: prediction,\n\u001b[0;32m   2697\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpresence_penalty\u001b[39m\u001b[38;5;124m\"\u001b[39m: presence_penalty,\n\u001b[0;32m   2698\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprompt_cache_key\u001b[39m\u001b[38;5;124m\"\u001b[39m: prompt_cache_key,\n\u001b[0;32m   2699\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprompt_cache_retention\u001b[39m\u001b[38;5;124m\"\u001b[39m: prompt_cache_retention,\n\u001b[0;32m   2700\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreasoning_effort\u001b[39m\u001b[38;5;124m\"\u001b[39m: reasoning_effort,\n\u001b[0;32m   2701\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse_format\u001b[39m\u001b[38;5;124m\"\u001b[39m: response_format,\n\u001b[0;32m   2702\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msafety_identifier\u001b[39m\u001b[38;5;124m\"\u001b[39m: safety_identifier,\n\u001b[0;32m   2703\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mseed\u001b[39m\u001b[38;5;124m\"\u001b[39m: seed,\n\u001b[0;32m   2704\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mservice_tier\u001b[39m\u001b[38;5;124m\"\u001b[39m: service_tier,\n\u001b[0;32m   2705\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstop\u001b[39m\u001b[38;5;124m\"\u001b[39m: stop,\n\u001b[0;32m   2706\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstore\u001b[39m\u001b[38;5;124m\"\u001b[39m: store,\n\u001b[0;32m   2707\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m: stream,\n\u001b[0;32m   2708\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream_options\u001b[39m\u001b[38;5;124m\"\u001b[39m: stream_options,\n\u001b[0;32m   2709\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtemperature\u001b[39m\u001b[38;5;124m\"\u001b[39m: temperature,\n\u001b[0;32m   2710\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtool_choice\u001b[39m\u001b[38;5;124m\"\u001b[39m: tool_choice,\n\u001b[0;32m   2711\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtools\u001b[39m\u001b[38;5;124m\"\u001b[39m: tools,\n\u001b[0;32m   2712\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtop_logprobs\u001b[39m\u001b[38;5;124m\"\u001b[39m: top_logprobs,\n\u001b[0;32m   2713\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtop_p\u001b[39m\u001b[38;5;124m\"\u001b[39m: top_p,\n\u001b[0;32m   2714\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muser\u001b[39m\u001b[38;5;124m\"\u001b[39m: user,\n\u001b[0;32m   2715\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mverbosity\u001b[39m\u001b[38;5;124m\"\u001b[39m: verbosity,\n\u001b[0;32m   2716\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mweb_search_options\u001b[39m\u001b[38;5;124m\"\u001b[39m: web_search_options,\n\u001b[0;32m   2717\u001b[0m         },\n\u001b[0;32m   2718\u001b[0m         completion_create_params\u001b[38;5;241m.\u001b[39mCompletionCreateParamsStreaming\n\u001b[0;32m   2719\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m stream\n\u001b[0;32m   2720\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m completion_create_params\u001b[38;5;241m.\u001b[39mCompletionCreateParamsNonStreaming,\n\u001b[0;32m   2721\u001b[0m     ),\n\u001b[0;32m   2722\u001b[0m     options\u001b[38;5;241m=\u001b[39mmake_request_options(\n\u001b[0;32m   2723\u001b[0m         extra_headers\u001b[38;5;241m=\u001b[39mextra_headers, extra_query\u001b[38;5;241m=\u001b[39mextra_query, extra_body\u001b[38;5;241m=\u001b[39mextra_body, timeout\u001b[38;5;241m=\u001b[39mtimeout\n\u001b[0;32m   2724\u001b[0m     ),\n\u001b[0;32m   2725\u001b[0m     cast_to\u001b[38;5;241m=\u001b[39mChatCompletion,\n\u001b[0;32m   2726\u001b[0m     stream\u001b[38;5;241m=\u001b[39mstream \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[0;32m   2727\u001b[0m     stream_cls\u001b[38;5;241m=\u001b[39mAsyncStream[ChatCompletionChunk],\n\u001b[0;32m   2728\u001b[0m )\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\openai\\_base_client.py:1881\u001b[0m, in \u001b[0;36mAsyncAPIClient.post\u001b[1;34m(self, path, cast_to, body, content, files, options, stream, stream_cls)\u001b[0m\n\u001b[0;32m   1878\u001b[0m opts \u001b[38;5;241m=\u001b[39m FinalRequestOptions\u001b[38;5;241m.\u001b[39mconstruct(\n\u001b[0;32m   1879\u001b[0m     method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpost\u001b[39m\u001b[38;5;124m\"\u001b[39m, url\u001b[38;5;241m=\u001b[39mpath, json_data\u001b[38;5;241m=\u001b[39mbody, content\u001b[38;5;241m=\u001b[39mcontent, files\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mawait\u001b[39;00m async_to_httpx_files(files), \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39moptions\n\u001b[0;32m   1880\u001b[0m )\n\u001b[1;32m-> 1881\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrequest(cast_to, opts, stream\u001b[38;5;241m=\u001b[39mstream, stream_cls\u001b[38;5;241m=\u001b[39mstream_cls)\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\openai\\_base_client.py:1666\u001b[0m, in \u001b[0;36mAsyncAPIClient.request\u001b[1;34m(self, cast_to, options, stream, stream_cls)\u001b[0m\n\u001b[0;32m   1665\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRe-raising status error\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m-> 1666\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_status_error_from_response(err\u001b[38;5;241m.\u001b[39mresponse) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m   1668\u001b[0m \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "\u001b[1;31mRateLimitError\u001b[0m: Error code: 429 - {'status': 429, 'message': 'Too Many Requests'}",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mRetryError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[19], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m ec_pred \u001b[38;5;241m=\u001b[39m \u001b[43mapi\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m      2\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpopulation\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mPatients with \u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mfin_condition\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m undergoing treatment with \u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mtreatements_eng\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m      3\u001b[0m \u001b[43m    \u001b[49m\u001b[43mintervention\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mtreatements_eng\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcomparator\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43moutcome\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m      6\u001b[0m \u001b[43m    \u001b[49m\u001b[43mllm\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetenv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mMODEL_NAME\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m      7\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcriteria\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mtitle_criteria\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mcontent_criteria\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m      8\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpapers\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mpapers\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;66;43;03m# make for the top-100 for demo\u001b[39;49;00m\n\u001b[0;32m      9\u001b[0m \u001b[43m)\u001b[49m\n\u001b[0;32m     10\u001b[0m ec_pred\n",
      "File \u001b[1;32m~\\GitHub\\TrialMind-SLR\\trialmind\\api.py:372\u001b[0m, in \u001b[0;36mLiteratureScreening.run\u001b[1;34m(self, population, intervention, comparator, outcome, papers, criteria, llm, batch_size)\u001b[0m\n\u001b[0;32m    370\u001b[0m \u001b[38;5;28;01mclass\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mPaperEvaluation\u001b[39;00m(BaseModel):\n\u001b[0;32m    371\u001b[0m     evaluations: conlist(Literal[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mYES\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNO\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUNCERTAIN\u001b[39m\u001b[38;5;124m'\u001b[39m], min_length\u001b[38;5;241m=\u001b[39mn_criteria, max_length\u001b[38;5;241m=\u001b[39mn_criteria) \u001b[38;5;241m=\u001b[39m Field(description\u001b[38;5;241m=\u001b[39m\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEvaluations for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mn_criteria\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m criteria, must be of length \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mn_criteria\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m--> 372\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[43mbatch_function_call_llm\u001b[49m\u001b[43m(\u001b[49m\u001b[43mLITERATURE_SCREENING_FC\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbatch_inputs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mPaperEvaluation\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mllm\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mllm\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbatch_size\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbatch_size\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    374\u001b[0m \u001b[38;5;66;03m# try to fix the predictions if not met the output format\u001b[39;00m\n\u001b[0;32m    375\u001b[0m parsed_outputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_outputs(outputs, n_criteria)\n",
      "File \u001b[1;32m~\\GitHub\\TrialMind-SLR\\trialmind\\llm.py:145\u001b[0m, in \u001b[0;36mbatch_function_call_llm\u001b[1;34m(prompt_template, batch_inputs, schema, llm, temperature, batch_size)\u001b[0m\n\u001b[0;32m    143\u001b[0m         results\u001b[38;5;241m.\u001b[39mextend(batch_results)\n\u001b[0;32m    144\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m--> 145\u001b[0m     results \u001b[38;5;241m=\u001b[39m \u001b[43mbatch_function_call_openai\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbatch_messages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mllm\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mllm\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtools\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtools\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemperature\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtemperature\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    146\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m results\n",
      "File \u001b[1;32m~\\GitHub\\TrialMind-SLR\\trialmind\\llm_utils\\openai_async.py:123\u001b[0m, in \u001b[0;36mbatch_function_call_openai\u001b[1;34m(batch_messages, llm, tools, temperature, thinking)\u001b[0m\n\u001b[0;32m    121\u001b[0m model \u001b[38;5;241m=\u001b[39m llm\u001b[38;5;66;03m#OPENAI_MODEL_NAME_MAP.get(llm)\u001b[39;00m\n\u001b[0;32m    122\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m model \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m--> 123\u001b[0m     results \u001b[38;5;241m=\u001b[39m \u001b[43m_async_execute\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    124\u001b[0m \u001b[43m        \u001b[49m\u001b[43masync_function\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mapply_function_call_async\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    125\u001b[0m \u001b[43m        \u001b[49m\u001b[43mclient\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43masync_openai_client\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    126\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    127\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmessages_list\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbatch_messages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    128\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtools\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtools\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    129\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtemperature\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtemperature\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m    130\u001b[0m \u001b[43m        \u001b[49m\u001b[43mseed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    131\u001b[0m \u001b[43m        \u001b[49m\u001b[43mthinking\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mthinking\u001b[49m\n\u001b[0;32m    132\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    133\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    134\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnknown llm: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mllm\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32m~\\GitHub\\TrialMind-SLR\\trialmind\\llm_utils\\openai_async.py:156\u001b[0m, in \u001b[0;36m_async_execute\u001b[1;34m(async_function, **kwargs)\u001b[0m\n\u001b[0;32m    154\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m ThreadPoolExecutor(\u001b[38;5;241m1\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m executor:\n\u001b[0;32m    155\u001b[0m         results \u001b[38;5;241m=\u001b[39m executor\u001b[38;5;241m.\u001b[39msubmit(\u001b[38;5;28;01mlambda\u001b[39;00m: asyncio\u001b[38;5;241m.\u001b[39mrun(async_function(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)))\n\u001b[1;32m--> 156\u001b[0m         results \u001b[38;5;241m=\u001b[39m \u001b[43mresults\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mresult\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    157\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m:\n\u001b[0;32m    158\u001b[0m     results \u001b[38;5;241m=\u001b[39m async_function(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\concurrent\\futures\\_base.py:458\u001b[0m, in \u001b[0;36mFuture.result\u001b[1;34m(self, timeout)\u001b[0m\n\u001b[0;32m    456\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m CancelledError()\n\u001b[0;32m    457\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_state \u001b[38;5;241m==\u001b[39m FINISHED:\n\u001b[1;32m--> 458\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m__get_result\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    459\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    460\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTimeoutError\u001b[39;00m()\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\concurrent\\futures\\_base.py:403\u001b[0m, in \u001b[0;36mFuture.__get_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    401\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception:\n\u001b[0;32m    402\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 403\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception\n\u001b[0;32m    404\u001b[0m     \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m    405\u001b[0m         \u001b[38;5;66;03m# Break a reference cycle with the exception in self._exception\u001b[39;00m\n\u001b[0;32m    406\u001b[0m         \u001b[38;5;28mself\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\concurrent\\futures\\thread.py:58\u001b[0m, in \u001b[0;36m_WorkItem.run\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     55\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[0;32m     57\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m---> 58\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfn(\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mkwargs)\n\u001b[0;32m     59\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[0;32m     60\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfuture\u001b[38;5;241m.\u001b[39mset_exception(exc)\n",
      "File \u001b[1;32m~\\GitHub\\TrialMind-SLR\\trialmind\\llm_utils\\openai_async.py:155\u001b[0m, in \u001b[0;36m_async_execute.<locals>.<lambda>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    153\u001b[0m     loop \u001b[38;5;241m=\u001b[39m asyncio\u001b[38;5;241m.\u001b[39mget_running_loop()\n\u001b[0;32m    154\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m ThreadPoolExecutor(\u001b[38;5;241m1\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m executor:\n\u001b[1;32m--> 155\u001b[0m         results \u001b[38;5;241m=\u001b[39m executor\u001b[38;5;241m.\u001b[39msubmit(\u001b[38;5;28;01mlambda\u001b[39;00m: \u001b[43masyncio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43masync_function\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[0;32m    156\u001b[0m         results \u001b[38;5;241m=\u001b[39m results\u001b[38;5;241m.\u001b[39mresult()\n\u001b[0;32m    157\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m:\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\asyncio\\runners.py:44\u001b[0m, in \u001b[0;36mrun\u001b[1;34m(main, debug)\u001b[0m\n\u001b[0;32m     42\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m debug \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m     43\u001b[0m         loop\u001b[38;5;241m.\u001b[39mset_debug(debug)\n\u001b[1;32m---> 44\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mloop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_until_complete\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmain\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     45\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m     46\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\asyncio\\base_events.py:649\u001b[0m, in \u001b[0;36mBaseEventLoop.run_until_complete\u001b[1;34m(self, future)\u001b[0m\n\u001b[0;32m    646\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m future\u001b[38;5;241m.\u001b[39mdone():\n\u001b[0;32m    647\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEvent loop stopped before Future completed.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m--> 649\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfuture\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mresult\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\GitHub\\TrialMind-SLR\\trialmind\\llm_utils\\openai_async.py:93\u001b[0m, in \u001b[0;36mapply_function_call_async\u001b[1;34m(client, model, messages_list, tools, **kwargs)\u001b[0m\n\u001b[0;32m     89\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     90\u001b[0m \u001b[38;5;124;03mApply the OpenAI API asynchronously to a list of messages using high-level asyncio APIs.\u001b[39;00m\n\u001b[0;32m     91\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     92\u001b[0m tasks \u001b[38;5;241m=\u001b[39m [api_function_call_single(client, model, messages, tools, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mfor\u001b[39;00m messages \u001b[38;5;129;01min\u001b[39;00m messages_list]\n\u001b[1;32m---> 93\u001b[0m results \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m asyncio\u001b[38;5;241m.\u001b[39mgather(\u001b[38;5;241m*\u001b[39mtasks)\n\u001b[0;32m     94\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m results\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\tenacity\\asyncio\\__init__.py:189\u001b[0m, in \u001b[0;36mAsyncRetrying.wraps.<locals>.async_wrapped\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    187\u001b[0m copy \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[0;32m    188\u001b[0m async_wrapped\u001b[38;5;241m.\u001b[39mstatistics \u001b[38;5;241m=\u001b[39m copy\u001b[38;5;241m.\u001b[39mstatistics  \u001b[38;5;66;03m# type: ignore[attr-defined]\u001b[39;00m\n\u001b[1;32m--> 189\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m copy(fn, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\tenacity\\asyncio\\__init__.py:111\u001b[0m, in \u001b[0;36mAsyncRetrying.__call__\u001b[1;34m(self, fn, *args, **kwargs)\u001b[0m\n\u001b[0;32m    109\u001b[0m retry_state \u001b[38;5;241m=\u001b[39m RetryCallState(retry_object\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m, fn\u001b[38;5;241m=\u001b[39mfn, args\u001b[38;5;241m=\u001b[39margs, kwargs\u001b[38;5;241m=\u001b[39mkwargs)\n\u001b[0;32m    110\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m--> 111\u001b[0m     do \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39miter(retry_state\u001b[38;5;241m=\u001b[39mretry_state)\n\u001b[0;32m    112\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(do, DoAttempt):\n\u001b[0;32m    113\u001b[0m         \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\tenacity\\asyncio\\__init__.py:153\u001b[0m, in \u001b[0;36mAsyncRetrying.iter\u001b[1;34m(self, retry_state)\u001b[0m\n\u001b[0;32m    151\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    152\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m action \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39miter_state\u001b[38;5;241m.\u001b[39mactions:\n\u001b[1;32m--> 153\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mawait\u001b[39;00m action(retry_state)\n\u001b[0;32m    154\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\tenacity\\_utils.py:99\u001b[0m, in \u001b[0;36mwrap_to_async_func.<locals>.inner\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     98\u001b[0m \u001b[38;5;28;01masync\u001b[39;00m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21minner\u001b[39m(\u001b[38;5;241m*\u001b[39margs: typing\u001b[38;5;241m.\u001b[39mAny, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: typing\u001b[38;5;241m.\u001b[39mAny) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m typing\u001b[38;5;241m.\u001b[39mAny:\n\u001b[1;32m---> 99\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m call(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mD:\\Programs\\Anaconda\\envs\\llms\\lib\\site-packages\\tenacity\\__init__.py:421\u001b[0m, in \u001b[0;36mBaseRetrying._post_stop_check_actions.<locals>.exc_check\u001b[1;34m(rs)\u001b[0m\n\u001b[0;32m    419\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreraise:\n\u001b[0;32m    420\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m retry_exc\u001b[38;5;241m.\u001b[39mreraise()\n\u001b[1;32m--> 421\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m retry_exc \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mfut\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mexception\u001b[39;00m()\n",
      "\u001b[1;31mRetryError\u001b[0m: RetryError[<Future at 0x27c8ee3a200 state=finished raised RateLimitError>]"
     ]
    }
   ],
   "source": [
    "ec_pred = api.run(\n",
    "    population = f\"Patients with {fin_condition} undergoing treatment with {treatements_eng[0]}\",\n",
    "    intervention = f\"{treatements_eng[0]}\",\n",
    "    comparator = \"\",\n",
    "    outcome = \"\",\n",
    "    llm = os.getenv(\"MODEL_NAME\"),\n",
    "    criteria = title_criteria + content_criteria,\n",
    "    papers = papers[:3], # make for the top-100 for demo\n",
    ")\n",
    "ec_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44cefd9e-0064-4e35-bc8a-b3bb9838fc4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c69ff7c4-4e91-4afc-bdae-ec1fbf094012",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/41268440/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/41213063/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40526090/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40211189/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/RESTful/pmcoa.cgi/BioC_xml/40369167/unicode?api_key=6892c4129cef143ff92d11533848d2e0d908\n"
     ]
    }
   ],
   "source": [
    "pmid_list = df_papers[0].PMID.values.tolist()\n",
    "#['41213063',#'26451310',]\n",
    "\n",
    "res = pmid2biocxml(pmid_list, api_key=pubmed_api_key)\n",
    "#print(res)\n",
    "res = [parse_bioc_xml(r) for r in res]\n",
    "# transform the parsed xml into paper content\n",
    "papers = []\n",
    "for parsed in res:\n",
    "    paper_content = []\n",
    "    for parsed_ in parsed[\"passage\"]:\n",
    "        paper_content.append(parsed_['content'])\n",
    "    paper_content = \"\\n\".join(paper_content)\n",
    "    papers.append(paper_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ac1dabb-884d-4bd7-9ce3-edd08f6c9e35",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1672, 1652, 27171, 2627, 46819]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[len(i) for i in papers]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6a9ab051-ac79-4050-93d9-7008b5074ed6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Case Report: HGF and NF1 mutations as putative bypass mechanisms of MET inhibitor resistance in hepatocellular carcinoma: a case study.',\n",
       " 'Background: Hepatocellular carcinoma (HCC) is a highly aggressive liver cancer with poor prognosis, often associated with resistance to treatment. MET amplification has been identified as a potential therapeutic target, but resistance to MET inhibitors, such as crizotinib, remains a significant challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic',\n",
       " 'challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,',\n",
       " 'analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically, a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts. Conclusion: Our',\n",
       " 'a missense mutation in HGF (p.G401A) was observed, which may enhance ligand-receptor binding, while an NF1 mutation (p.M546L) may permit sustained MAPK and PI3K activation despite MET inhibition. These observations are preliminary and require validation in larger patient cohorts. Conclusion: Our findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to',\n",
       " 'Our findings suggest that acquired resistance to MET inhibitors in MET-amplified HCC may involve clonal evolution and activation of compensatory signaling pathways. These insights highlight the need for dynamic molecular surveillance and the development of strategies targeting multiple pathways to overcome resistance and improve patient outcomes.']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "splited = split_text_into_chunks(papers[1], chunk_size=600,\n",
    "                                             chunk_overlap=300)\n",
    "splited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ac23d300-fc62-407f-8265-409a50f8a216",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "False\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[{'name': 'Crizotinib treatment effectiveness',\n",
       "   'value': 'initial tumor regression followed by disease progression',\n",
       "   'source_id': [2],\n",
       "   'cited_blocks': ['challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,']}],\n",
       " [{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'More sensitive to low-risk group',\n",
       "   'source_id': [5],\n",
       "   'cited_blocks': ['performed in risk groups. Expression of signature genes in CRC cells was examined using quantitative reverse transcription polymerase chain reaction (qRT-PCR). There were 10 CRC-related RRAGs found. Prognostic models were built with RRAGs. Based on immune infiltration analysis, low-risk populations showed significantly greater levels of immune infiltration (P P ARHGEF17 mutation rate was 6%. Medications such as CGP-082996, Dasatinib, Erlotinib, and Salubrinal were more sensitive to high-risk group, whereas drugs such as FTI-277, DMOG, and Crizotinib were more sensitive to low-risk group.']}],\n",
       " [{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'NP',\n",
       "   'source_id': [],\n",
       "   'cited_blocks': []}],\n",
       " [{'name': 'main_results_crizotinib',\n",
       "   'value': 'no objective responses observed in RASMT advanced CRC patients',\n",
       "   'source_id': [7],\n",
       "   'cited_blocks': ['with low baselineÂ\\xa0KRASMT allele frequency. CONCLUSIONS: Combination binimetinib/crizotinib showed a poor tolerability with no objective responses observed in RASMT advanced CRC patients. EudraCT-Number: 2014-000463 -â\\x80\\x8940 (20/06/2014: A Sequential Phase I study of MEK1/2 inhibitors PD- 0325901 or Binimetinib combined with cMET inhibitor Crizotinib in RAS Mutant and RAS Wild Type with aberrant c-MET).']}],\n",
       " [{'name': 'Crizotinib_effectiveness',\n",
       "   'value': 'selectively inhibited the growth of ARID1A-deficient CRC cells',\n",
       "   'source_id': [80, 142],\n",
       "   'cited_blocks': ['validated the synthetic lethal effect of the c-MET inhibitor PHA-665752 (PHA) in two ARID1A knockout (ARID1A-KO) clones, where PHA selectively inhibited the growth of ARID1A-KO HCT116 cells (Fig.Â\\xa01D). Moreover, two FDA-approved c-MET inhibitors, crizotinib, and cabozantinib, also showed synthetic lethality in ARID1A-KO HCT116 clones (Fig.Â\\xa01E, F). These results suggest that c-MET is a potential synthetic lethal target for ARID1A in CRC cells.',\n",
       "    'synthetic lethality in ARID1A-deficient CRC. Notably, FDA-approved c-MET inhibitors, such as crizotinib and cabozantinib, have demonstrated enhanced cytotoxicity in ARID1A-deficient cancers, facilitating the translation of these approved drugs for repurposing in ARID1A-targeted precision cancer medicine.']}]]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "api = StudyCharacteristicsExtraction()\n",
    "extracted = api.run(\n",
    "    papers=papers,\n",
    "    fields=['Results, string, main results of the study regarding Crizotinib, usually about treatment effectiveness'],#api.DEFAULT_FIELDS,\n",
    "    llm='Qwen/Qwen3-32B',#\"openai-gpt-4o\",\n",
    "    chunk_size=600,\n",
    "    chunk_overlap=300,\n",
    "    thinking=False,\n",
    ")\n",
    "extracted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "dc85b6da-8fe9-4484-8430-e55e239b2525",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[{'name': 'Crizotinib effectiveness',\n",
       "   'value': 'More sensitive to low-risk group',\n",
       "   'source_id': [4],\n",
       "   'cited_blocks': ['performed in risk groups. Expression of signature genes in CRC cells was examined using quantitative reverse transcription polymerase chain reaction (qRT-PCR). There were 10 CRC-related RRAGs found. Prognostic models were built with RRAGs. Based on immune infiltration analysis, low-risk populations showed significantly greater levels of immune infiltration (P P ARHGEF17 mutation rate was 6%. Medications such as CGP-082996, Dasatinib, Erlotinib, and Salubrinal were more sensitive to high-risk group, whereas drugs such as FTI-277, DMOG, and Crizotinib were more sensitive to low-risk group.']}],\n",
       " [{'name': 'Crizotinib treatment effectiveness',\n",
       "   'value': 'initial tumor regression followed by disease progression',\n",
       "   'source_id': [1],\n",
       "   'cited_blocks': ['challenge. This study aims to explore the molecular mechanisms underlying resistance to MET inhibitors in MET-amplified HCC. Methods: We present a case of advanced HCC in a patient with MET amplification treated with crizotinib. After initial tumor regression, disease progression occurred. Genetic analysis using next-generation sequencing (NGS) was performed on biopsy samples taken before and after progression to identify mutations associated with resistance. Results: NGS revealed the loss of MET amplification and identified HGF and NF1 mutations as potential bypass mechanisms. Specifically,']}]]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "80601546-e32e-4b65-b837-b5dec0abdfc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "llm_reasoning = ChatOpenAI(\n",
    "    base_url=BASE_URL,\n",
    "    api_key=HUGGINGFACE_HUB_TOKEN,\n",
    "    model=MODEL_NAME_HF,\n",
    "    streaming=True,\n",
    "    temperature=0,\n",
    "    #http_client=httpx_client\n",
    "    http_async_client=httpx.AsyncClient(verify=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "87efd589-f082-4326-bc29-42ab4cd01c9a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<think>\n",
      "\n",
      "</think>\n",
      "\n",
      "```json\n",
      "[\n",
      "    {\n",
      "        \"name\": \"Crizotinib effectiveness\",\n",
      "        \"value\": \"More sensitive to low-risk group\",\n",
      "        \"source_id\": [3]\n",
      "    }\n",
      "]\n",
      "```\n"
     ]
    }
   ],
   "source": [
    "prompt = ChatPromptTemplate.from_messages([\n",
    "    (\"user\", '/no_think {text}'),\n",
    "    # (\"ai\", \"Good.\")\n",
    "])\n",
    "\n",
    "chain = prompt | llm_reasoning\n",
    "response = await chain.ainvoke({\"text\": ff})\n",
    "\n",
    "#fin_condition = response.content.strip('<think>\\n\\n</think>\\n\\n')\n",
    "print(response.content)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:llms]",
   "language": "python",
   "name": "conda-env-llms-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
